---
title: "R Notebook"
output: html_notebook
---

```{r include = FALSE, echo = FALSE, warning=FALSE}
library(tidyverse)
library(EnhancedVolcano)
library(gplots)
library(ggvenn)
library(DESeq2)
library(goseq)
library(GO.db)
library(reshape2)
library(GOsummaries)
library(cowplot)
library(broom)
library(gridExtra)
library(Rgraphviz)
library(broom)
library(ggrepel)
library(gridExtra)
library(topGO)
library(ggpmisc)
library(GOxploreR)
library(UpSetR)
library(ComplexHeatmap)
library(ggplotify)
library(corrplot)
library(vegan)
library(cluster)
library(GGally)
library(wesanderson)
```

```{r}
load("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/topGO_plots_outline.Rdata")

```

```{r}
norm_expr_trt <- 
  norm_expr[,rownames(subset(colData, Treatment == "treatment"))] %>%
  t(.) %>%
  merge(colData, ., by = 0) %>%
  column_to_rownames(., var = "Row.names") %>%
  dplyr::select(., c(Gen_Group, Treatment, starts_with("SCA-6"))) %>%
  pivot_longer(cols = starts_with("SCA-6"), values_to = "expression", names_to = "Gene") %>%
  group_by(Gen_Group, Gene) %>%
  summarise(mean_expression = mean(expression)) %>%
  pivot_wider(id_cols = Gene, values_from = mean_expression, names_from = Gen_Group) %>%
  column_to_rownames(., var = "Gene") %>%
  add_column(., .before = 1, Full = rowMeans(.)) %>%
  rownames_to_column(., var = "Gene")

norm_expr_ctl <- 
  norm_expr[,rownames(subset(colData, Treatment == "control"))] %>%
  t(.) %>%
  merge(colData, ., by = 0) %>%
  column_to_rownames(., var = "Row.names") %>%
  dplyr::select(., c(Gen_Group, Treatment, starts_with("SCA-6"))) %>%
  pivot_longer(cols = starts_with("SCA-6"), values_to = "expression", names_to = "Gene") %>%
  group_by(Gen_Group, Gene) %>%
  summarise(mean_expression = mean(expression)) %>%
  pivot_wider(id_cols = Gene, values_from = mean_expression, names_from = Gen_Group) %>%
  column_to_rownames(., var = "Gene") %>%
  add_column(., .before = 1, Full = rowMeans(.)) %>%
  rownames_to_column(., var = "Gene")

norm_expr_phe <- 
  norm_expr[,rownames(subset(colData, Phenotype == "R"))] %>%
  t(.) %>%
  merge(colData, ., by = 0) %>%
  column_to_rownames(., var = "Row.names") %>%
  dplyr::select(., c(Gen_Group, Treatment, starts_with("SCA-6"))) %>%
  pivot_longer(cols = starts_with("SCA-6"), values_to = "expression", names_to = "Gene") %>%
  group_by(Gen_Group, Gene) %>%
  summarise(mean_expression = mean(expression)) %>%
  pivot_wider(id_cols = Gene, values_from = mean_expression, names_from = Gen_Group) %>%
  column_to_rownames(., var = "Gene") %>%
  add_column(., .before = 1, Full = rowMeans(.)) %>%
  rownames_to_column(., var = "Gene")


```


# TREATMENT EFFECT--Full Model--GO & KEGG Enrichment
## Setup TopGO functions
```{r}
# Function to convery SCA-6 GO terms to list that can be used by TopGO
readMappings <- function(file, sep = "\t", IDsep = ",") {
  a <- read.delim(file = file, header = FALSE,
                  quote = "", sep = sep, colClasses = "character")
  
  ## a bit of preprocesing to get it in a nicer form
  map <- a[, 2]
  names(map) <- gsub(" ", "", a[, 1]) ## trim the spaces
  
  ## split the IDs
  return(lapply(map, function(x) gsub(" ", "", strsplit(x, split = IDsep)[[1]])))
}

# How do you want to define significance--here it is genes with padj < 0.05, could choose LFC cutoff
topDiffGenes05 <- function(allScore) {return(allScore < 0.05)}

# Load geneID2GO file using readMappings function 
geneIDGO <- readMappings(file ="~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/SCA6_GO.txt", sep = '\t', IDsep = ';')

# Load GO IDs
got <- read.table("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/SCA6_GO.txt", 
                  header = TRUE, 
                  sep = '\t', 
                  stringsAsFactors = FALSE) %>%
  separate_rows(., GO_ID, sep = ';') %>%
  mutate(., Gene = gsub("\\.1", "", .$Gene))

```

# Pull top n genes from each contrast and genetic group
```{r}
genetic_group <- c("iquitos","nanay", "guiana", "maranon")
full_mods <- c("TmainLFC", "PmainLFC", "PTinteractLFC")
n <- 1000

TmainLFC_guiana <- rownames_to_column(TmainLFC_guiana, var = "Gene")
TmainLFC_iquitos <- rownames_to_column(TmainLFC_iquitos, var = "Gene")
TmainLFC_maranon <- rownames_to_column(TmainLFC_maranon, var = "Gene")
TmainLFC_nanay <- rownames_to_column(TmainLFC_nanay, var = "Gene")

PmainLFC_guiana <- rownames_to_column(PmainLFC_guiana, var = "Gene")
PmainLFC_iquitos <- rownames_to_column(PmainLFC_iquitos, var = "Gene")
PmainLFC_maranon <- rownames_to_column(PmainLFC_maranon, var = "Gene")
PmainLFC_nanay <- rownames_to_column(PmainLFC_nanay, var = "Gene")

PTinteractLFC_guiana <- rownames_to_column(PTinteractLFC_guiana, var = "Gene")
PTinteractLFC_iquitos <- rownames_to_column(PTinteractLFC_iquitos, var = "Gene")
PTinteractLFC_maranon <- rownames_to_column(PTinteractLFC_maranon , var = "Gene")
PTinteractLFC_nanay <- rownames_to_column(PTinteractLFC_nanay, var = "Gene")

for (i in 1:length(full_mods)){
  df <- full_mods[i]
  
  # Cut according to lfc threshold
  for (grp in genetic_group){
    # What is the name of the df?
    dfname_orig <- paste0(df, "_", grp)
    # Access the value of that name from the global env
    var.value <- as.list(environment())[[dfname_orig]]
    var.value <- data.frame(var.value)
    var.value$abs <- abs(var.value$log2FoldChange)
    # Name of set object
    set_name <- paste0(dfname_orig, "_lfc", n)
    # Extract names of significant genes--lfc threshold
    set_lfc <- var.value[order(var.value$abs, decreasing = TRUE),] %>%
      .[1:n,]
    # Merge with annotation file
    #set_lfc <- inner_join(set_lfc, annot, by = "Gene")
    # Assign name to object
    assign(set_name, set_lfc)
  }
  
  # Genes that show up as significant, padj < 0.05
  for (grp in genetic_group){
    # What is the name of the df?
    dfname_orig <- paste0(df,"_", grp)
    # Access the value of that name from the global env
    var.value <- as.list(environment())[[dfname_orig]]
    # Name of set object
    set_name <- paste0(df, "_", grp, "_sig")
    # Extract names of significant genes--pval threshold
    set_padj <- subset(var.value, padj < 0.05) 
    
    if (nrow(set_padj) > 0){
      # Merge with annotation file 
      set_padj <- inner_join(set_padj, annot, by = "Gene")
      # Assign name to object
      assign(set_name, set_padj)
    } else{
      assign(set_name, set_padj)
    }
  }
}




```

# Additive effects for each genetic group
```{r}
# Additive effects--i.e. treatment + phenotype
df1 <- "TmainLFC"
df2 <- "PmainLFC"

PTadditiveLFC <- inner_join(subset(TMainLFC, padj < 0.05), 
                            subset(PMainLFC, padj < 0.05),
                            by = "Gene")

for (grp in genetic_group){
  
    # What is the name of the df?
    dfname1 <- paste0(df1,"_", grp, "_lfc", n)
    dfname2 <- paste0(df2,"_", grp, "_lfc", n)
    
    # Access the value of that name from the global env
    var.value1 <- as.list(environment())[[dfname1]]
    var.value1 <- data.frame(var.value1) 
    
    var.value2 <- as.list(environment())[[dfname2]]
    var.value2 <- data.frame(var.value2) 
    
    # Name the new df
    set_name <- paste0("PTadditiveLFC_", grp, "_lfc", n)
    
    # Intersection between phenotype and treatment for additive effects--add annotations
    addit <- inner_join(var.value1, var.value2, by = "Gene")
    addit <- inner_join(addit, annot, by = "Gene")
  
    # Assign name to object
    assign(set_name, addit)
}

# Combine all LFC estimates for TREATMENT
TmainLFC_all_pop <- 
  rbind(TmainLFC_guiana %>% add_column(.,Population = rep("Guiana", nrow(.))),
           TmainLFC_iquitos %>% add_column(.,Population = rep("Iquitos", nrow(.))),
           TmainLFC_maranon %>% add_column(.,Population = rep("Marañón", nrow(.))),
           TmainLFC_nanay %>% add_column(.,Population = rep("Nanay", nrow(.))))

TmainLFC_all_pop_lfc1000 <- 
  rbind(TmainLFC_guiana_lfc1000 %>% add_column(.,Population = rep("Guiana", nrow(.))),
           TmainLFC_iquitos_lfc1000 %>% add_column(.,Population = rep("Iquitos", nrow(.))),
           TmainLFC_maranon_lfc1000 %>% add_column(.,Population = rep("Marañón", nrow(.))),
           TmainLFC_nanay_lfc1000 %>% add_column(.,Population = rep("Nanay", nrow(.))))

# Combind all LFC estimates for PHENOTYPE
PmainLFC_all_pop <- 
  rbind(PmainLFC_guiana %>% add_column(.,Population = rep("Guiana", nrow(.))),
           PmainLFC_iquitos %>% add_column(.,Population = rep("Iquitos", nrow(.))),
           PmainLFC_maranon %>% add_column(.,Population = rep("Marañón", nrow(.))),
           PmainLFC_nanay %>% add_column(.,Population = rep("Nanay", nrow(.))))

PmainLFC_all_pop_lfc1000 <- 
  rbind(PmainLFC_guiana_lfc1000 %>% add_column(.,Population = rep("Guiana", nrow(.))),
           PmainLFC_iquitos_lfc1000 %>% add_column(.,Population = rep("Iquitos", nrow(.))),
           PmainLFC_maranon_lfc1000 %>% add_column(.,Population = rep("Marañón", nrow(.))),
           PmainLFC_nanay_lfc1000 %>% add_column(.,Population = rep("Nanay", nrow(.))))



```

################################################
## Run GO enrichments -- genetic group models ##
################################################
```{r include = FALSE}
#############################################
### Treat, Pheno, Interactive Effects--GO ###
#############################################
df1 <- "TmainLFC"
df2 <- "PmainLFC"
genetic_group <- c("iquitos", "maranon", "guiana", "nanay")
full_mods <- c("TmainLFC", "PmainLFC")
n <- 1000
lfc_df <- unlist(lapply(genetic_group, function(x) {paste0(full_mods, "_", x, "_lfc", n)}))

# Load geneID2GO file using readMappings function 
geneIDGO <- readMappings(file ="~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/SCA6_GO.txt", sep = '\t', IDsep = ';')

for (df1 in lfc_df){ 
  print(df1)
  # Pull out the dataframes
  var.value1 <- as.list(environment())[[df1]]
  
  # Need the full dataframe (not subsetted) for allExprGenes
  df2<- gsub("_lfc.*", "", df1)
  var.value2 <- as.list(environment())[[df2]]
  
  # What LFC cutoff should we use 
  cutoff <- min(abs(var.value1$log2FoldChange))
  
  # What gnes are we using 
  allExprGenes <- var.value2 %>% 
  #rownames_to_column(., var = "Gene") %>%
  dplyr::select(., c(Gene, log2FoldChange)) %>%
  deframe(.)

  # How do you want to define significance--here it is genes with padj < 0.05, could choose LFC cutoff
  topDiffGenesLFC <- function(allScore) {return(abs(allScore) > cutoff)}
  
  # Create the topGO object--BP
  GOdata <- new("topGOdata",
                    description = "Simple session", ontology = "BP",
                    allGenes = allExprGenes, geneSel = topDiffGenesLFC,
                    nodeSize = 10,
                    annot = annFUN.gene2GO, 
                    gene2GO = geneIDGO)
  
  # Run the enrichment tests, correct p-values 
  resultClassic <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  topGO_bp <- GenTable(GOdata, pval = resultClassic, topNodes = length(resultClassic@score))
  topGO_bp$padj <- p.adjust(topGO_bp$pval, method = "fdr")
  topGO_bp$ontology <- rep("BP", nrow(topGO_bp))
  
  # Create the topGO object--MF
  GOdata <- new("topGOdata",
                    description = "Simple session", ontology = "MF",
                    allGenes = allExprGenes, geneSel = topDiffGenesLFC,
                    nodeSize = 10,
                    annot = annFUN.gene2GO, 
                    gene2GO = geneIDGO)
  
  # Run the enrichment tests, correct p-values 
  resultClassic <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  topGO_mf <- GenTable(GOdata, pval = resultClassic, topNodes = length(resultClassic@score))
  topGO_mf$padj <- p.adjust(topGO_mf$pval, method = "fdr")
  topGO_mf$ontology <- rep("MF", nrow(topGO_mf))
  
  tmp_comb <- rbind(topGO_bp, topGO_mf)
  
  dfname <- paste0(df1, "_topGOenrichment_combOntology")
  assign(dfname, tmp_comb)
  
}
```

#######################################
## Run GO enrichments -- full models ##
#######################################

```{r include = FALSE}
############################
### Treatment Effect--GO ###
############################
ont <- c("BP", "MF")
for (i in ont){
  allExprGenes <- data.frame(TMainLFC) %>% 
  #rownames_to_column(., var = "Gene") %>%
  dplyr::select(., c(Gene, padj)) %>%
  deframe(.)

  GOdata <- new("topGOdata",
                      description = "Simple session", ontology = i,
                      allGenes = allExprGenes, geneSel = topDiffGenes05,
                      nodeSize = 10,
                      annot = annFUN.gene2GO, 
                      gene2GO = geneIDGO)
  
  resultClassic <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  TMainLFC_topGO <- GenTable(GOdata, 
                             pval = resultClassic,
                             topNodes = length(resultClassic@score))
  TMainLFC_topGO$padj <- p.adjust(TMainLFC_topGO$pval, method = "fdr")
  TMainLFC_topGO$ontology <- rep(i, nrow(TMainLFC_topGO))
  
  name <- paste0("TMainLFC_topGO_", i)
  assign(name, TMainLFC_topGO)
}

TMainLFC_topGO <- rbind(TMainLFC_topGO_BP, TMainLFC_topGO_MF)

############################
### Phenotype Effect--GO ###
############################
for (i in ont){
  allExprGenes <- data.frame(PMainLFC) %>% 
  #rownames_to_column(., var = "Gene") %>%
  dplyr::select(., c(Gene, padj)) %>%
  deframe(.)

  GOdata <- new("topGOdata",
                      description = "Simple session", ontology = i,
                      allGenes = allExprGenes, geneSel = topDiffGenes05,
                      nodeSize = 10,
                      annot = annFUN.gene2GO, 
                      gene2GO = geneIDGO)
  
  resultClassic <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  PMainLFC_topGO <- GenTable(GOdata, 
                             pval = resultClassic, 
                             topNodes = length(resultClassic@score))
  PMainLFC_topGO$padj <- p.adjust(PMainLFC_topGO$pval, method = "fdr")
  PMainLFC_topGO$ontology <- rep(i, nrow(PMainLFC_topGO))
  
  name <- paste0("PMainLFC_topGO_", i)
  assign(name, PMainLFC_topGO)
}

PMainLFC_topGO <- rbind(PMainLFC_topGO_BP, PMainLFC_topGO_MF)

###########################
### Additive Effect--GO ###
###########################
for (i in ont){
  allExprGenes <- data.frame(PTadditiveLFC) %>% 
  #rownames_to_column(., var = "Gene") %>%
  dplyr::select(., c(Gene, padj.x)) %>%
  deframe(.)

  GOdata <- new("topGOdata",
                      description = "Simple session", ontology = i,
                      allGenes = allExprGenes, geneSel = topDiffGenes05,
                      nodeSize = 10,
                      annot = annFUN.gene2GO, 
                      gene2GO = geneIDGO)
  
  resultClassic <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  PMainLFC_topGO <- GenTable(GOdata, 
                             pval = resultClassic, 
                             topNodes = length(resultClassic@score))
  PMainLFC_topGO$padj <- p.adjust(PMainLFC_topGO$pval, method = "fdr")
  PMainLFC_topGO$ontology <- rep(i, nrow(PMainLFC_topGO))
  
  name <- paste0("PMainLFC_topGO_", i)
  assign(name, PMainLFC_topGO)
}

PTadditiveLFC_topGO <- rbind(PTadditiveLFC_topGO_BP, PTadditiveLFC_topGO_MF)

```

#### Rank using GOxploreR
```{r}
go_dfs <- list(TmainLFC_guiana_lfc1000_topGOenrichment_combOntology, 
            TmainLFC_iquitos_lfc1000_topGOenrichment_combOntology,
            TmainLFC_maranon_lfc1000_topGOenrichment_combOntology,
            TmainLFC_nanay_lfc1000_topGOenrichment_combOntology,
            PmainLFC_guiana_lfc1000_topGOenrichment_combOntology,
            PmainLFC_iquitos_lfc1000_topGOenrichment_combOntology,
            PmainLFC_maranon_lfc1000_topGOenrichment_combOntology,
            PmainLFC_nanay_lfc1000_topGOenrichment_combOntology)

names(go_dfs) <- c("TmainLFC_guiana_lfc1000_topGOenrichment_combOntology", 
            "TmainLFC_iquitos_lfc1000_topGOenrichment_combOntology",
            "TmainLFC_maranon_lfc1000_topGOenrichment_combOntology",
            "TmainLFC_nanay_lfc1000_topGOenrichment_combOntology",
            "PmainLFC_guiana_lfc1000_topGOenrichment_combOntology",
            "PmainLFC_iquitos_lfc1000_topGOenrichment_combOntology",
            "PmainLFC_maranon_lfc1000_topGOenrichment_combOntology",
            "PmainLFC_nanay_lfc1000_topGOenrichment_combOntology")

for (k in 1:length(go_dfs)){
  terms <- go_dfs[[k]]
  terms <- subset(terms, ontology == "BP" & padj < 0.05)
  slim_id <- prioritizedGOTerms(as.character(terms$GO.ID), organism = "Cress", sp=TRUE, domain = "BP")$HF
  slimGOTerms <- terms[terms$GO.ID %in% slim_id,]
  
  nm <- gsub("_topGOenrichment_combOntology", "", names(go_dfs)[k])
  set_name <- paste0(nm, "_slimGOTerms")
  print(set_name)
  assign(set_name, slimGOTerms)
}

# Full model--treatment
terms <- subset(TMainLFC_topGO, ontology == "BP" & padj < 0.05)$GO.ID
slimGOTerms <- prioritizedGOTerms(terms, organism = "Cress", sp=TRUE, domain = "BP")$HF
TMainLFC_topGO_slimGOTerms <- TMainLFC_topGO[TMainLFC_topGO$GO.ID %in% slimGOTerms,]

# Full model -- phenotype
terms <- subset(PMainLFC_topGO, ontology == "BP" & padj < 0.05)
slimGOTerms <- prioritizedGOTerms(terms, organism = "Cress", sp=TRUE, domain = "BP")$HF
PMainLFC_topGO_slimGOTerms <- PMainLFC_topGO[PMainLFC_topGO$GO.ID %in% slimGOTerms,]




```

##############
## PLOTTING ##
##############
## Calculations and plots for figure displaying divergent expression profiles across groups
#### PLOT -- Upset plot of DEG overlaps for each genetic group
```{r}
lt <- list(Guiana = as.character(TmainLFC_guiana_lfc1000$Gene),
           Iquitos = as.character(TmainLFC_iquitos_lfc1000$Gene),
           Maranon = as.character(TmainLFC_maranon_lfc1000$Gene),
           Nanay = as.character(TmainLFC_nanay_lfc1000$Gene))
treat_df <- list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

lt <- list(Guiana = as.character(PmainLFC_guiana_lfc1000$Gene),
           Iquitos = as.character(PmainLFC_iquitos_lfc1000$Gene),
           Maranon = as.character(PmainLFC_maranon_lfc1000$Gene),
           Nanay = as.character(PmainLFC_nanay_lfc1000$Gene))
pheno_df <- list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

lt <- list(Guiana = as.character(PTadditiveLFC_guiana_lfc1000$Gene),
           Iquitos = as.character(PTadditiveLFC_iquitos_lfc1000$Gene),
           Maranon = as.character(PTadditiveLFC_maranon_lfc1000$Gene),
           Nanay = as.character(PTadditiveLFC_nanay_lfc1000$Gene))
addit_df <- list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig5_upsetPlots_degOverlaps.pdf", width = 12)

 upset(treat_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Genes (Largest |LFC|)",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))
 
upset(pheno_df, main.bar.color = "black",
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  keep.order = TRUE,
  sets.x.label = "Genes (Largest |LFC|)",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))

upset(addit_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Genes (Largest |LFC|)",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))

dev.off()

```

#### PLOT -- DEG correlation Matrix of Log2 Fold Changes
```{r}

trt_log2FC_allPop <- 
  inner_join(TMainLFC[,c("Gene", "log2FoldChange")], TmainLFC_guiana[,c("Gene", "log2FoldChange")],
                by = "Gene") %>%
  inner_join(.,TmainLFC_iquitos[,c("Gene", "log2FoldChange")], 
             by = "Gene") %>%
  inner_join(.,TmainLFC_maranon[,c("Gene", "log2FoldChange")],
             by = "Gene") %>%
  inner_join(.,TmainLFC_nanay[,c("Gene", "log2FoldChange")],
             by = "Gene") %>%
  setNames(., nm = c("Gene", "Full", "Guiana", "Iquitos", "Marañón", "Nanay")) %>% 
  dplyr::select(-Full) %>% 
  column_to_rownames(var = "Gene")



phe_log2FC_allPop <- 
  inner_join(PMainLFC[,c("Gene", "log2FoldChange")], PmainLFC_guiana[,c("Gene", "log2FoldChange")],
                by = "Gene") %>%
  inner_join(.,PmainLFC_iquitos[,c("Gene", "log2FoldChange")], 
             by = "Gene") %>%
  inner_join(.,PmainLFC_maranon[,c("Gene", "log2FoldChange")],
             by = "Gene") %>%
  inner_join(.,PmainLFC_nanay[,c("Gene", "log2FoldChange")],
             by = "Gene") %>%
  setNames(., nm = c("Gene", "Full", "Guiana", "Iquitos", "Marañón", "Nanay")) %>% 
  dplyr::select(-Full) %>% 
  column_to_rownames(var = "Gene")

trt_go_effect <-
  trt_log2FC_allPop %>%
  rownames_to_column(var = "Gene") %>%
  .[.$Gene %in% unique(TmainLFC_all_pop_lfc1000$Gene), ] %>%
  #dplyr::select(-Full) %>%
  remove_rownames() %>%
  column_to_rownames(var = "Gene")

pheno_go_effect <-
  phe_log2FC_allPop %>%
  rownames_to_column(var = "Gene") %>%
  .[.$Gene %in% unique(PmainLFC_all_pop_lfc1000$Gene), ] %>%
  #dplyr::select(-Full) %>%
  remove_rownames() %>%
  column_to_rownames(var = "Gene")

#
#
# trt_go_expr <- norm_expr_trt[norm_expr_trt$Gene %in% unique(TmainLFC_all_pop_lfc1000$Gene),] %>%
#   dplyr::select(-Full) %>%
#   remove_rownames() %>%
#   column_to_rownames(var = "Gene")
# 
# pheno_go_expr <- norm_expr_phe[norm_expr_phe$Gene %in% unique(PmainLFC_all_pop_lfc1000$Gene),] %>%
#   dplyr::select(-Full) %>%
#   remove_rownames() %>%
#   column_to_rownames(var = "Gene")
#
#

# For ALL genes
trt_pmat <- cor.mtest(trt_log2FC_allPop, alternative = "two.sided", method = "spearman")
trt_pmat$p[lower.tri(trt_pmat$p)] <- 1
phe_pmat <- cor.mtest(phe_log2FC_allPop, alternative = "two.sided", method = "spearman")
phe_pmat$p[lower.tri(phe_pmat$p)] <- 1

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig5_log2FC_corrMatrix_all.pdf", height = 25, width = 30)
par(mfrow=c(1,2))

# LFC correlation matrices for ALL genes -- Treatment
corrplot.mixed(cor(trt_log2FC_allPop, method = "spearman"), p.mat = trt_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)


# LFC orrelation matrices for ALL genes -- Phenotype
corrplot.mixed(cor(phe_log2FC_allPop, method = "spearman"), p.mat = phe_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3, tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

dev.off()

# For only DE genes
trt_pmat <- cor.mtest(trt_go_effect, alternative = "two.sided", method = "spearman")
trt_pmat$p[lower.tri(trt_pmat$p)] <- 1
phe_pmat <- cor.mtest(pheno_go_effect, alternative = "two.sided", method = "spearman")
phe_pmat$p[lower.tri(phe_pmat$p)] <- 1

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig5_log2FC_corrMatrix_onlyDE.pdf", height = 25, width = 30)
par(mfrow=c(2,1))

# LFC correlation matrices for ALL genes -- Treatment
corrplot.mixed(cor(trt_go_effect, method = "spearman"), p.mat = trt_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)


# LFC orrelation matrices for ALL genes -- Phenotype
corrplot.mixed(cor(pheno_go_effect, method = "spearman"), p.mat = phe_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3, tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

dev.off()


# trt_log2FC_allPop <- melt(x, id.vars = c("Gene", "Full"), value.name = "log2FoldChange", variable.name = "Population")
# my.formula <- y ~ x
# pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_log2FC_corr.pdf",
#     height = 10,
#     width = 25)
# ggplot(trt_log2FC_allPop, aes(x = Full, y = log2FoldChange)) + 
#   geom_point(alpha = 0.2) + 
#   geom_smooth(method = "lm", color = "black", linetype = 2) +
#   facet_grid(~ Population) +
#   theme_minimal_grid() +
#   labs(x = "log2FoldChange Full Model", y = "log2FoldChange Population Models") +
#   stat_poly_eq(formula = my.formula, 
#                          aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
#                          parse = TRUE)
# dev.off()

```

#### PLOT -- Upset plot of orthogroup overlaps for each genetic group
```{r}
# Read in orthogroup data frame
orthos <- read.table("../SCA6.proteins.both.37Gv1.0.bestOrthos", header = TRUE, sep = '\t') %>%
  setNames(nm = c("Gene", "Orthogroup"))

# Merge treatment, phenotype, and additive dataframes with orhogroups dataframe
ortho_tMainLFC <- inner_join(orthos, TmainLFC_all_pop_lfc1000, by = "Gene")
ortho_pMainLFC <- inner_join(orthos, PmainLFC_all_pop_lfc1000, by = "Gene")
ortho_additLFC <- 
  data.frame(
    rbind(
      inner_join(orthos, PTadditiveLFC_guiana_lfc1000, by = "Gene") %>% 
      add_column(Population = rep("Guiana", nrow(.))),
    inner_join(orthos, PTadditiveLFC_iquitos_lfc1000, by = "Gene") %>% 
      add_column(Population = rep("Iquitos", nrow(.))),
    inner_join(orthos, PTadditiveLFC_maranon_lfc1000, by = "Gene") %>% 
      add_column(Population = rep("Marañón", nrow(.))),
    inner_join(orthos, PTadditiveLFC_nanay_lfc1000, by = "Gene") %>% 
      add_column(Population = rep("Nanay", nrow(.)))
    )
  )

# Treatment
lt <- 
  list(
    Guiana = unique(as.character(subset(ortho_tMainLFC, Population == "Guiana")$Orthogroup)),
    Iquitos = unique(as.character(subset(ortho_tMainLFC, Population == "Iquitos")$Orthogroup)),
    Maranon = unique(as.character(subset(ortho_tMainLFC, Population == "Marañón")$Orthogroup)),
    Nanay = unique(as.character(subset(ortho_tMainLFC, Population == "Nanay")$Orthogroup)))

ortho_tMainLFC_df <- 
  list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

# Phenotype
lt <- 
  list(
    Guiana = unique(as.character(subset(ortho_pMainLFC, Population == "Guiana")$Orthogroup)),
    Iquitos = unique(as.character(subset(ortho_pMainLFC, Population == "Iquitos")$Orthogroup)),
    Maranon = unique(as.character(subset(ortho_pMainLFC, Population == "Marañón")$Orthogroup)),
    Nanay = unique(as.character(subset(ortho_pMainLFC, Population == "Nanay")$Orthogroup)))

ortho_pMainLFC_df <- 
  list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

# Additive
lt <- list(Guiana = unique(as.character(subset(ortho_additLFC, Population == "Guiana")$Orthogroup)),
           Iquitos = unique(as.character(subset(ortho_additLFC, Population == "Iquitos")$Orthogroup)),
           Maranon = unique(as.character(subset(ortho_additLFC, Population == "Marañón")$Orthogroup)),
           Nanay = unique(as.character(subset(ortho_additLFC, Population == "Nanay")$Orthogroup)))

ortho_additLFC_df <- list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/Fig5_upsetPlots_orthogroupOverlaps.pdf", width = 12) 
# Treatment
upset(ortho_tMainLFC_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Orthogroups",
  mainbar.y.label = "OGs in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))

# Phenotype
upset(ortho_pMainLFC_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Orthogroups",
  mainbar.y.label = "OGs in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))

# Additive
upset(ortho_additLFC_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Orthogroups",
  mainbar.y.label = "OGs in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))
dev.off()


```

#### Correlation matrices of orthogroup LFCs
```{r}
ortho_tMainLFC_cor <- 
  TmainLFC_all_pop %>%
  dplyr::select(c(Gene, log2FoldChange, Population)) %>%
  .[.$Gene %in% unique(TmainLFC_all_pop_lfc1000$Gene),] %>% 
  inner_join(orthos, by = "Gene") %>% 
  group_by(Population, Orthogroup) %>% 
  summarise(mLFC = mean(log2FoldChange)) %>% 
  pivot_wider(names_from = Population, values_from = mLFC) %>% 
  mutate(Orthogroup = paste0("OG", Orthogroup)) %>%
  column_to_rownames(var = "Orthogroup")

#ortho_tMainLFC_cor <- cor(ortho_tMainLFC_cor, method = "spearman")

ortho_pMainLFC_cor <- 
  PmainLFC_all_pop %>%
  dplyr::select(c(Gene, log2FoldChange, Population)) %>%
  .[.$Gene %in% unique(PmainLFC_all_pop_lfc1000$Gene),] %>% 
  inner_join(orthos, by = "Gene") %>% 
  group_by(Population, Orthogroup) %>% 
  summarise(mLFC = mean(log2FoldChange)) %>% 
  pivot_wider(names_from = Population, values_from = mLFC) %>% 
  mutate(Orthogroup = paste0("OG", Orthogroup)) %>%
  column_to_rownames(var = "Orthogroup")

#ortho_pMainLFC_cor <- cor(ortho_pMainLFC_cor, method = "spearman")


ortho_trt_pmat <- cor.mtest(ortho_tMainLFC_cor, alternative = "two.sided", method = "spearman")
ortho_trt_pmat$p[lower.tri(ortho_trt_pmat$p)] <- 1
ortho_phe_pmat <- cor.mtest(ortho_pMainLFC_cor, alternative = "two.sided", method = "spearman")
ortho_phe_pmat$p[lower.tri(ortho_phe_pmat$p)] <- 1

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/Fig5_log2FC_corrMatrix_orthogroups.pdf", 
    height = 25, width = 30)
par(mfrow=c(1,2))

# LFC correlation matrices for TOP genes -- Treatment
corrplot.mixed(cor(ortho_tMainLFC_cor, method = "spearman"), p.mat = ortho_trt_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)


# LFC orrelation matrices for TOP genes -- Phenotype
corrplot.mixed(cor(ortho_pMainLFC_cor, method = "spearman"), p.mat = ortho_phe_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3, tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

dev.off()

```

#### Correlation matrices for Arabidopsis LFC -- Winkelmüller, Entila, Anver et al. (2021)
##### From their supplemental data: https://academic.oup.com/plcell/article/33/6/1863/6159620#283679405
```{r}
ath_lfc <- read.csv("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/arath_geenExprEvol_flg22_LFC_winkelmuller2021.csv", header = TRUE)

rel_lfc <- ath_lfc %>% 
  dplyr::select(c("Arabidopsis_thaliana", "Ath", "Cru", "Chi", "Esa")) %>%
  column_to_rownames(var = "Arabidopsis_thaliana")

acc_lfc <- ath_lfc %>%
  dplyr::select(c("Arabidopsis_thaliana", "Ath", "Can0", "Gy0", "Kn0", "Kon", "No0")) %>%
  column_to_rownames(var = "Arabidopsis_thaliana")
  
  

rel_lfc_pmat <- cor.mtest(rel_lfc, alternative = "two.sided", method = "spearman")
rel_lfc_pmat$p[lower.tri(rel_lfc_pmat$p)] <- 1
acc_lfc_pmat <- cor.mtest(acc_lfc, alternative = "two.sided", method = "spearman")
acc_lfc_pmat$p[lower.tri(acc_lfc_pmat$p)] <- 1

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/arath_log2FC_corrMatrix_oneToOneOrthos.pdf", 
    height = 25, width = 30)
par(mfrow=c(1,2))

# LFC orrelation matrices for RELATIVES of A thaliana
corrplot.mixed(cor(rel_lfc, method = "spearman"), p.mat = rel_lfc_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3, tl.col = c("black", "black", "black", "black","black", "black"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

# LFC orrelation matrices for ACCESSIONS of A thaliana
corrplot.mixed(cor(acc_lfc, method = "spearman"), p.mat = acc_lfc_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3, tl.col = c("black", "black", "black", "black","black", "black"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)
dev.off()


```

#### DEG overlap when using FDR-adjusted p-values
```{r}
# Function to calculate SEM
sem <- function(x) sqrt(var(x)/length(x))

# Overlap when subsetting by FDR-adjusted p-values -- TREATMENT 
fdr_deg_treat <- 
  TmainLFC_all_pop %>%
  group_by(Population) %>% 
  filter(padj < 0.05)

lt <- list(
  as.character(subset(fdr_deg_treat, Population == "Guiana")$Gene),
  as.character(subset(fdr_deg_treat, Population == "Iquitos")$Gene),
  as.character(subset(fdr_deg_treat, Population == "Marañón")$Gene),
  as.character(subset(fdr_deg_treat, Population == "Nanay")$Gene)
)
names(lt) <- c("Guiana", "Iquitos", "Marañón", "Nanay")

fdr_trt_propUniq <- c(
  "Guiana"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Guiana') / length(lt[["Guiana"]]),
  "Iquitos"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Iquitos') / length(lt[["Iquitos"]]),
  "Marañón"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Marañón') / length(lt[["Marañón"]]),
  "Nanay"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Nanay') / length(lt[["Nanay"]])
)

fdr_trt_propUniq
mean(fdr_trt_propUniq) * 100
sem(fdr_trt_propUniq) * 100

# Overlap when subsetting by FDR-adjusted p-values -- PHENOTYPE 
fdr_deg_pheno <- 
  PmainLFC_all_pop %>%
  group_by(Population) %>% 
  filter(padj < 0.05)

lt <- list(
  as.character(subset(fdr_deg_pheno, Population == "Guiana")$Gene),
  as.character(subset(fdr_deg_pheno, Population == "Iquitos")$Gene),
  as.character(subset(fdr_deg_pheno, Population == "Marañón")$Gene),
  as.character(subset(fdr_deg_pheno, Population == "Nanay")$Gene)
)
names(lt) <- c("Guiana", "Iquitos", "Marañón", "Nanay")

fdr_phe_propUniq <- c(
  "Guiana"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Guiana') / length(lt[["Guiana"]]),
  "Iquitos"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Iquitos') / length(lt[["Iquitos"]]),
  "Marañón"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Marañón') / length(lt[["Marañón"]]),
  "Nanay"= length(attr(venn(lt, show.plot = FALSE), "intersections")$'Nanay') / length(lt[["Nanay"]])
)

fdr_phe_propUniq
mean(fdr_phe_propUniq) * 100
sem(fdr_phe_propUniq) * 100


```




## Calculation and plots for figure showing divergent effect sizes for enriched GO terms
### Calculate median LFC for slimmed GO terms --  model for each group
```{r}
# Extract genes in each GO category, but only those that are differentially expressed
gene_go_long <- read.table(file = "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/SCA6_GO.txt", sep = '\t', header = TRUE, stringsAsFactors = FALSE) %>%
  separate_rows(., GO_ID, sep = ';') %>%
  setNames(., nm = c("Gene", "GO.ID")) 

go_dfs <- c("TmainLFC_guiana_lfc1000_slimGOTerms", 
               "TmainLFC_iquitos_lfc1000_slimGOTerms",
               "TmainLFC_maranon_lfc1000_slimGOTerms",
               "TmainLFC_nanay_lfc1000_slimGOTerms", 
               "PmainLFC_guiana_lfc1000_slimGOTerms",
               "PmainLFC_iquitos_lfc1000_slimGOTerms",
               "PmainLFC_maranon_lfc1000_slimGOTerms",
               "PmainLFC_nanay_lfc1000_slimGOTerms")

# LOSING A LOT OF GO TERMS HERE, NEED TO FIGURE OUT WHY
for (i in go_dfs){
  # Pull out the dataframes--slim GO terms
  var.value1 <- as.list(environment())[[i]]
  
  # Data frame containing the gene IDs and GO termms
  i2<- gsub("_slimGOTerms", "", i)
  var.value2 <- as.list(environment())[[i2]]
  
  # Merge GO slim terms with gene IDs + GO terms, then with df containing LFCs
  tmp <- inner_join(var.value1, gene_go_long, by = "GO.ID") %>%
    inner_join(., var.value2, by = "Gene")
  
  set_name <- paste0(i, "_LFC")
  assign(set_name, tmp)
  
}
```

### Calculate median LFC for slimmed GO terms --  full model
```{r}
# TMainLFC_topGO_slimGOTerms_LFC <- inner_join(TMainLFC_topGO_slimGOTerms, gene_go_long, by = "GO.ID") %>%
#   inner_join(., subset(TMainLFC, padj < 0.05), by = "Gene") %>%
#   mutate(., Population = rep("Full", nrow(.)))
# PMainLFC_topGO_slimGOTerms_LFC <- inner_join(PMainLFC_topGO_slimGOTerms, gene_go_long, by = "GO.ID")%>%
#   inner_join(., subset(PMainLFC, padj < 0.05), by = "Gene") %>%
#   mutate(., Population = rep("Full", nrow(.)))
# 
# # Median fold change for all significant genes in a given GO category -- full mode, treatment
# TMainLFC_topGO_slimGOTerms_medLFC <- 
#   TMainLFC_topGO_slimGOTerms_LFC %>%
#   group_by(GO.ID) %>%
#   tally() %>%
#   filter(., n > 10) %>%
#   inner_join(., TMainLFC_topGO_slimGOTerms_LFC, by = "GO.ID") %>%
#   group_by(GO.ID) %>%
#   summarize(medlog2LFC = median(log2FoldChange), 
#             meanlog2FC = mean(log2FoldChange),
#             medFC = 2^median(log2FoldChange),
#             meanFC = 2^mean(log2FoldChange)) %>%
#   arrange(., desc(medlog2LFC)) %>%
#   head(., n = 10) %>%
#   inner_join(., TMainLFC_topGO_slimGOTerms_LFC, by = "GO.ID") %>%
#   distinct(GO.ID, .keep_all = TRUE) %>%
#   .[,1:12] 
# 
# # TMainLFC_topGO_slimGOTerms_LFC <- inner_join(TMainLFC_topGO_slimGOTerms_LFC,
# #                                              TMainLFC_topGO_slimGOTerms_medLFC,
# #                                              by = "GO.ID")
# 
# # Median fold change for all significant genes in a given GO category -- full mode, phenotype
# # No GO categories with 10 genes or fewer
# # Note: THERE ARE NO GO TERMS SIG ENRICHED IN THE FULL PHENOTYPE MODEL
# PMainLFC_topGO_slimGOTerms_medLFC <- 
#   PMainLFC_topGO_slimGOTerms_LFC %>%
#   group_by(GO.ID) %>%
#   tally() %>%
#   filter(., n > 10) %>%
#   inner_join(., PMainLFC_topGO_slimGOTerms_LFC, by = "GO.ID") %>%
#   group_by(GO.ID) %>%
#   summarize(medlog2LFC = median(log2FoldChange), 
#             meanlog2FC = mean(log2FoldChange),
#             medFC = 2^median(log2FoldChange),
#             meanFC = 2^mean(log2FoldChange)) %>%
#   arrange(., desc(medlog2LFC)) %>%
#   head(., n = 10) %>%
#   inner_join(., PMainLFC_topGO_slimGOTerms_LFC, by = "GO.ID") %>%
#   distinct(GO.ID, .keep_all = TRUE) %>%
#   .[,1:12]
#   
# 
# PMainLFC_topGO_slimGOTerms_LFC <- inner_join(PMainLFC_topGO_slimGOTerms_LFC,
#                                              PMainLFC_topGO_slimGOTerms_medLFC, by = "GO.ID")
# 


```

### Extract LFC for all genes in TMainLFC enriched GO terms--check effect sizes
```{r}
# # Effect size for top GO terms from the full model, i.e. are the genes from the GO terms that are enriched in the full model behaving in the same way across genetic groups
# 
# # Guiana
# gu_topGO_effectSize <- TMainLFC_topGO_slimGOTerms_LFC[TMainLFC_topGO_slimGOTerms_LFC$GO.ID %in% TMainLFC_topGO_slimGOTerms_medLFC$GO.ID,] %>%
#   dplyr::select(., c(GO.ID, Term, Gene)) %>%
#   inner_join(., TmainLFC_guiana, by = "Gene") %>% 
#   # group_by(GO.ID) %>%
#   # summarize(medlog2LFC = median(log2FoldChange), 
#   #           meanlog2FC = mean(log2FoldChange),
#   #           medFC = 2^median(log2FoldChange),
#   #           meanFC = 2^mean(log2FoldChange)) %>%
#   #setNames(., nm = c("GO.ID", paste0("guiana_", colnames(.)[-1]) )) %>%
#   mutate(., Population = rep("Guiana", nrow(.))) 
# 
# # Iquitos
# iq_topGO_effectSize <- TMainLFC_topGO_slimGOTerms_LFC[TMainLFC_topGO_slimGOTerms_LFC$GO.ID %in% TMainLFC_topGO_slimGOTerms_medLFC$GO.ID,] %>%
#   dplyr::select(., c(GO.ID, Term, Gene)) %>%
#   inner_join(., TmainLFC_iquitos, by = "Gene") %>% 
#   # group_by(GO.ID) %>%
#   # summarize(medlog2LFC = median(log2FoldChange), 
#   #           meanlog2FC = mean(log2FoldChange),
#   #           medFC = 2^median(log2FoldChange),
#   #           meanFC = 2^mean(log2FoldChange)) %>%
#   #setNames(., nm = c("GO.ID", paste0("iquitos_", colnames(.)[-1]) )) %>%
#   mutate(., Population = rep("Iquitos", nrow(.))) 
# 
# # Maranon
# ma_topGO_effectSize <- TMainLFC_topGO_slimGOTerms_LFC[TMainLFC_topGO_slimGOTerms_LFC$GO.ID %in% TMainLFC_topGO_slimGOTerms_medLFC$GO.ID,] %>%
#   dplyr::select(., c(GO.ID, Term, Gene)) %>%
#   inner_join(., TmainLFC_maranon, by = "Gene") %>% 
#   # group_by(GO.ID) %>%
#   # summarize(medlog2LFC = median(log2FoldChange), 
#   #           meanlog2FC = mean(log2FoldChange),
#   #           medFC = 2^median(log2FoldChange),
#   #           meanFC = 2^mean(log2FoldChange)) %>%
#   #setNames(., nm = c("GO.ID", paste0("maranon_", colnames(.)[-1]) )) %>%
#   mutate(., Population = rep("Marañón", nrow(.))) 
# 
# # Nanay
# na_topGO_effectSize <- TMainLFC_topGO_slimGOTerms_LFC[TMainLFC_topGO_slimGOTerms_LFC$GO.ID %in% TMainLFC_topGO_slimGOTerms_medLFC$GO.ID,] %>%
#   dplyr::select(., c(GO.ID, Term, Gene)) %>%
#   inner_join(., TmainLFC_nanay, by = "Gene") %>% 
#   # group_by(GO.ID) %>%
#   # summarize(medlog2LFC = median(log2FoldChange), 
#   #           meanlog2FC = mean(log2FoldChange),
#   #           medFC = 2^median(log2FoldChange),
#   #           meanFC = 2^mean(log2FoldChange)) %>%
#   #setNames(., nm = c("GO.ID", paste0("nanay_", colnames(.)[-1]) )) %>%
#   mutate(., Population = rep("Nanay", nrow(.))) 
# 
# # all_go_lfc <- inner_join(gu_topGO_effectSize, iq_topGO_effectSize, by = "GO.ID") %>%
# #   inner_join(., ma_topGO_effectSize, by = "GO.ID") %>%
# #   inner_join(., na_topGO_effectSize, by = "GO.ID") %>%
# #   inner_join(., TMainLFC_topGO_slimGOTerms_medLFC, by = "GO.ID")
# all_go_lfc <-
#   TMainLFC_topGO_slimGOTerms_LFC[TMainLFC_topGO_slimGOTerms_LFC$GO.ID %in% TMainLFC_topGO_slimGOTerms_medLFC$GO.ID,] %>%
#   .[,c(1:2,9:16)] %>%
#   setNames(., nm = c(colnames(.)[-c(9:10)], "padj", "Population")) %>%
#   rbind(.,
#         gu_topGO_effectSize,
#         iq_topGO_effectSize,
#         ma_topGO_effectSize,
#         na_topGO_effectSize) %>%
#   mutate(., Contrast = rep("Treatment", nrow(.)))
# 
# fac_cols <- data.frame(unique(all_go_lfc$GO.ID)) %>%
#   mutate(cols = rep(c(1:5), times = (nrow(.) / 5))) %>%
#   setNames(., nm = c("GO.ID", "Columns"))
# 
# all_go_lfc <- inner_join(all_go_lfc, fac_cols, by = "GO.ID")
# 
# # all_go_lfc <- rbind(TMainLFC_topGO_slimGOTerms_LFC[,c(1:2,9:14)],
# #                     gu_topGO_effectSize,
# #                     iq_topGO_effectSize, 
# #                     ma_topGO_effectSize, 
# #                     na_topGO_effectSize) %>%
# #   mutate(., Contrast = rep("Treatment", nrow(.)))
#                    

```

#### PLOT -- Boxplot of effect size for enriched GO terms from TMainLFC
```{r}
# Only plotting effect size for each go for the TREATMENT effetc--P
# None of the enriched GO terms in the PHENOTYPE effect had > 10 genes
# pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_BinGO_enrichedGOEffectSize.pdf",
#     height = 25,
#     width = 50)
# all_go_lfc %>%
#   ggplot() + 
#   geom_point(aes(y=tolower(Term),
#              x=1,
#              size=medFC,
#              color = Population), 
#              ) + 
#   scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
#   scale_x_continuous(limits = c(1,1), breaks = 0) +
#   facet_grid(Contrast~Population) + 
#   guides(color = "none") + 
#   labs(size = "Median Fold Change") +
#   theme(strip.placement = "outside",
#         plot.title = element_text(hjust = 0.5),
#         axis.title.y = element_blank(),
#         axis.text.x = element_blank(),
#         axis.ticks.y = element_blank(),
#         axis.ticks.x = element_blank(),
#         strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
#         strip.text.x = element_text(size = 30, face = "bold"),
#         strip.text.y = element_text(size = 30, face = "bold"),
#         axis.text.y = element_text(size = 30, face = "bold"),
#         legend.text = element_text(size = 30),
#         legend.title = element_text(size = 30),
#         legend.position = "bottom") +
#   scale_size_continuous(range = c(10,25))
# dev.off()
# 
# 
# pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_boxplot_enrichedGOEffectSize.pdf",
#     height = 35,
#     width = 20)
# # Boxplots for effect size rather than dotplot
# all_go_lfc %>%
#   mutate(FoldChange = 2^(log2FoldChange)) %>%
#   ggplot(aes(x=Population, y=FoldChange, color = Population)) +
#   #geom_violin(width = 0.75, alpha = 0.7) +
#   geom_boxplot(lwd = 2) +
#   geom_jitter(size = 3, alpha = 0.7) +
#   scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
#   labs(x = NULL, y = "Fold Change") +
#   scale_y_continuous(limits = c(-5,5)) +
#   facet_wrap(~ Term, ncol = 2, nrow = 5) +
#   theme(strip.placement =  "inside",
#         plot.title = element_text(hjust = 0.5),
#         axis.text.x = element_blank(),
#         axis.ticks.x = element_blank(),
#         strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
#         strip.text.x = element_text(size = 30, face = "bold"),
#         strip.text.y = element_text(size = 30, face = "bold"),
#         axis.text.y = element_text(size = 30, face = "bold"),
#         axis.title.y = element_text(size = 30, face = "bold"),
#         legend.text = element_text(size = 30),
#         legend.title = element_text(size = 30),
#         legend.position = "bottom")
# dev.off()
# 
# 

```

#### PLOT -- BinGO overlap between GO terms in each genetic group -- BinGO + LFC
```{r}
# PLOT -- BinGO GO overlaps + effect size
# Function to calculate median and mean LFC for each enriched GO term in each genetic group
go_dfs <- c("TmainLFC_guiana_lfc1000_slimGOTerms_LFC",
               "TmainLFC_iquitos_lfc1000_slimGOTerms_LFC",
               "TmainLFC_maranon_lfc1000_slimGOTerms_LFC",
               "TmainLFC_nanay_lfc1000_slimGOTerms_LFC",
               "PmainLFC_guiana_lfc1000_slimGOTerms_LFC",
               "PmainLFC_iquitos_lfc1000_slimGOTerms_LFC",
               "PmainLFC_maranon_lfc1000_slimGOTerms_LFC",
               "PmainLFC_nanay_lfc1000_slimGOTerms_LFC")

for (i in go_dfs){
  # Pull out the dataframes--slim GO terms
  var.value1 <- as.list(environment())[[i]]
  
  # Gen group
  i2 <- gsub("[TP]mainLFC_", "", i) %>%
    gsub("_lfc1000_slimGOTerms_LFC", "", .) %>%
    str_to_title(.)
  
  # Summarize median LFC for each enriched GO term in each genetic group
  tmp <-
    var.value1 %>%
    group_by(GO.ID) %>%
    tally() %>%
    filter(., n > 10) %>%
    inner_join(., var.value1, by = "GO.ID") %>%
    group_by(GO.ID) %>%
    summarise(medlog2LFC = median(log2FoldChange), 
              meanlog2FC = mean(log2FoldChange),
              medFC = 2^median(log2FoldChange),
              meanFC = 2^mean(log2FoldChange)) %>%
    arrange(., desc(medlog2LFC)) %>%
    inner_join(., var.value1, by = "GO.ID") %>%
    distinct(GO.ID, .keep_all = TRUE) %>%
    .[,1:12] %>%
    mutate(., Population = rep(i2, nrow(.)))

  # Data frame containing the gene IDs and GO termms
  set_name<- gsub("_LFC", "_medLFC", i)
  
  # Assign set name to tmp
  assign(set_name, tmp)
  
}

# TMainLFC_topGO_slimGOTerms_medLFC$Population <- 
#   rep("Full", nrow(TMainLFC_topGO_slimGOTerms_medLFC))
  
trt_goOverlaps <- 
  rbind(TmainLFC_guiana_lfc1000_slimGOTerms_medLFC,
        TmainLFC_iquitos_lfc1000_slimGOTerms_medLFC,
        TmainLFC_maranon_lfc1000_slimGOTerms_medLFC,
        TmainLFC_nanay_lfc1000_slimGOTerms_medLFC) %>%
  mutate(., Contrast = rep("Treatment", nrow(.))) %>%
  mutate(., Population = ifelse(.$Population == "Maranon", "Marañón", .$Population))

phe_goOverlaps <- 
  rbind(PmainLFC_guiana_lfc1000_slimGOTerms_medLFC,
        PmainLFC_iquitos_lfc1000_slimGOTerms_medLFC,
        PmainLFC_maranon_lfc1000_slimGOTerms_medLFC,
        PmainLFC_nanay_lfc1000_slimGOTerms_medLFC) %>%
  mutate(., Contrast = rep("Phenotype", nrow(.))) %>%
  mutate(., Population = ifelse(.$Population == "Maranon", "Marañón", .$Population))

# Use ReviGO to reduce redundancy in the enriched GO terms--i.e. cluster GO terms by similarity
# http://revigo.irb.hr/?JobID=1397531132
# Arabidopsis thaliana, Lin method 
# Will use rrvgo in the future--but need R 4.0 and didn't want to break my pipelines
trt_goOverlaps %>%
  dplyr::select(., GO.ID) %>%
  unique() %>%
  mutate(GO.ID = gsub("GO:", "", .$GO.ID)) %>%
  write.table(., 
            "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_treatment.txt",
            sep = '\t',
            quote = FALSE, 
            row.names = FALSE,
            col.names = FALSE)

phe_goOverlaps %>%
  dplyr::select(., GO.ID) %>%
  unique() %>%
  mutate(GO.ID = gsub("GO:", "", .$GO.ID)) %>%
  write.table(., 
            "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_phenotype.txt",
            sep = '\t',
            quote = FALSE, 
            row.names = FALSE,
            col.names = FALSE)

# What are the terms associated with out GO.IDs
go_terms <- rbind(trt_goOverlaps, phe_goOverlaps) %>%
  dplyr::select(c(GO.ID, Term)) %>%
  unique()

# What GO IDs were merged--treatment
grouped_go_trt <- trt_goOverlaps_revigo_lfc <- 
  read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_treatment_Revigo.csv", sep = ',', ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>% 
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               TmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               TmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               TmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               TmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID") %>% 
  subset(GO.ID != Grouped_GO.ID) %>% 
  dplyr::select(c(GO.ID, Grouped_GO.ID)) %>% 
  inner_join(go_terms, by = c("Grouped_GO.ID" = "GO.ID")) %>%
  distinct()

# Plot median LFC for each enriched GO term (after GOexploreR and REVIGO filtering)
trt_goOverlaps_revigo_lfc <- 
  read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_treatment_Revigo.csv", sep = ',', ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>% 
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               TmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               TmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               TmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               TmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID") %>%
  group_by(Population, Grouped_GO.ID) %>%
  summarise(medLFC = abs(median(log2FoldChange))) %>%
  #rename(., Grouped_GO.ID = "GO.ID") %>%
  inner_join(., go_terms, by = c("Grouped_GO.ID" = "GO.ID")) %>%
  add_column(., Contrast = rep("Treatment", nrow(.)))

p1 <-
  ggplot(trt_goOverlaps_revigo_lfc) + 
  geom_point(aes(y=tolower(Term),
             x=1,
             size=2^(medLFC),
             color = Population), 
             ) + 
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
  scale_x_continuous(limits = c(1,1), breaks = 0) +
  facet_grid(Contrast~Population) + 
  guides(color = "none") + 
  labs(size = "Median Fold Change", x = NULL, y = NULL)+ 
  theme(
    strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
    strip.placement = "outside",
    axis.title.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.ticks.x = element_blank(),
    axis.text.y = element_text(size = 13),
    strip.text = element_text(size = 13),
    legend.text = element_text(size = 13),
    legend.position = "bottom",
    legend.justification = "center",
    legend.direction = "horizontal"
  )


# What GO IDs were merged -- Phenotype
grouped_go_phe <- 
  read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_phenotype_Revigo.csv", sep = ',', ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>% 
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               PmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               PmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               PmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               PmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID") %>%
  subset(GO.ID != Grouped_GO.ID) %>% 
  dplyr::select(c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(go_terms, by = c("Grouped_GO.ID" = "GO.ID")) %>%
  distinct()

# Plot median LFC for each enriched GO term (after GOexploreR and REVIGO filtering)
phe_goOverlaps_revigo_lfc <- 
  read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_phenotype_Revigo.csv", sep = ',', ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>% 
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               PmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               PmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               PmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               PmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID") %>%
  group_by(Population, Grouped_GO.ID) %>%
  summarise(medLFC = abs(median(log2FoldChange))) %>%
  #rename(., Grouped_GO.ID = "GO.ID") %>%
  inner_join(., go_terms, by = c("Grouped_GO.ID" = "GO.ID")) %>%
  add_column(., Contrast = rep("Phenotype", nrow(.)))

p2 <-
  ggplot(phe_goOverlaps_revigo_lfc) + 
  geom_point(aes(y=tolower(Term),
             x=1,
             size=2^(medLFC),
             color = Population), 
             ) + 
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
  scale_x_continuous(limits = c(1,1), breaks = 0) +
  facet_grid(Contrast~Population) + 
  guides(color = "none") + 
  labs(size = "Median Fold Change", x = NULL, y = NULL)+ 
  theme(
    strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
    strip.placement = "outside",
    axis.title.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.ticks.x = element_blank(),
    axis.text.y = element_text(size = 13),
    strip.text = element_text(size = 13),
    legend.text = element_text(size = 13),
    legend.position = "bottom",
    legend.justification = "center",
    legend.direction = "horizontal"
  )

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_goOverlaps.pdf", width = 15, height = 13)
plot_grid(p1, p2, nrow = 1, ncol = 2)
dev.off()

```

#### PLOT -- Overlap between GO terms in each genetic group -- upset plot
```{r}
# PLOT -- GO overlaps with upset plot
# Treatment
lt <- list(
  as.character(subset(trt_goOverlaps_revigo_lfc, Population == "Guiana")$Grouped_GO.ID),
  as.character(subset(trt_goOverlaps_revigo_lfc, Population == "Iquitos")$Grouped_GO.ID),
  as.character(subset(trt_goOverlaps_revigo_lfc, Population == "Marañón")$Grouped_GO.ID),
  as.character(subset(trt_goOverlaps_revigo_lfc, Population == "Nanay")$Grouped_GO.ID)
)

treat_df_go <- list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

# Phenotype
lt <- list(
  as.character(subset(phe_goOverlaps_revigo_lfc, Population == "Guiana")$Grouped_GO.ID),
  as.character(subset(phe_goOverlaps_revigo_lfc, Population == "Iquitos")$Grouped_GO.ID),
  as.character(subset(phe_goOverlaps_revigo_lfc, Population == "Marañón")$Grouped_GO.ID),
  as.character(subset(phe_goOverlaps_revigo_lfc, Population == "Nanay")$Grouped_GO.ID)
)
pheno_df_go <- list_to_matrix(lt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))


pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_upsetPlots_goOverlaps.pdf", width = 12)

upset(treat_df_go, main.bar.color = "black",
      queries = list(
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "GO Terms",
  mainbar.y.label = "GO Terms in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2))

 
upset(pheno_df_go, main.bar.color = "black",
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  queries = list(
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  keep.order = TRUE,
  sets.x.label = "GO Terms",
  mainbar.y.label = "GO Terms in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2))

dev.off()

# upset(addit_df, main.bar.color = "black",
#       queries = list(
#   list(query = intersects, params = list("Full"), color = "black", active = T),
#   list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
#   list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
#   list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
#   list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
#   #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
#   keep.order = TRUE,
#   sets.x.label = "Gene Sets (Largest log2FoldChange)",
#   mainbar.y.label = "Genes in Each Intersection",
#   point.size = 20,
#   line.size = 10, 
#   text.scale = 15)
```

#### PLOT -- LFC for conserved GO terms
```{r}
# What GO terms are found in all n populations
n <- 4
shared_goTerms_trt <- 
  treat_df_go %>% 
  add_column(shared = rowSums(.))  %>%
  filter(shared >= n) %>% 
  rownames(.)

shared_goTerms_phe <- 
  pheno_df_go %>% 
  add_column(shared = rowSums(.)) %>% 
  filter(shared >= n) %>% 
  rownames(.)


# LFC for top shared GO terms
# Make sure if GO terms have been grouped that genes are placed in the appropriate GO term

# Treatment
shared_goTerms_trt_LFC <- 
  read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_treatment_Revigo.csv", sep = ',' ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>%
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               TmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               TmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               TmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               TmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID")  %>%
  rename(Term = "Original_Term") %>%
  inner_join(go_terms, by = c("Grouped_GO.ID" = "GO.ID")) %>% 
  rename(Term = "Grouped_Term") %>%
  .[.$Grouped_GO.ID %in% shared_goTerms_trt,]


# Phenotype
shared_goTerms_phe_LFC <- 
  read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/NSF_modelSelection/go_overlaps_allPop_phenotype_revigo.csv", sep = ',' ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>%
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               PmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               PmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               PmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               PmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID")  %>%
  rename(Term = "Original_Term") %>%
  inner_join(go_terms, by = c("Grouped_GO.ID" = "GO.ID")) %>% 
  rename(Term = "Grouped_Term") %>% 
  .[.$Grouped_GO.ID %in% shared_goTerms_phe,]


# Calculate the # population-specific vs shared genes for each population x go category
# Treatment
genetic_group <- c("Guiana","Iquitos","Marañón","Nanay")
percSharedUniq_go_trt <- data.frame()
percSharedUniq_go_phe <- data.frame()
for (t in unique(shared_goTerms_trt_LFC$Grouped_Term)){
  
  tmp <- subset(shared_goTerms_trt_LFC, Grouped_Term == t)
  id <- unique(tmp$Grouped_GO.ID)
  
    for(g in genetic_group){
      
      uniq <- attr(venn(list(
        Guiana = as.character(subset(tmp, Population == "Guiana")$Gene),
        Iquitos = as.character(subset(tmp, Population == "Iquitos")$Gene),
        Marañón = as.character(subset(tmp, Population == "Marañón")$Gene),
        Nanay = as.character(subset(tmp, Population == "Nanay")$Gene)), 
        show.plot = FALSE),"intersections")
      uniq_gen <- as.numeric(length(uniq[[g]]))
      tot_gen <- nrow(subset(tmp, Population == g))
      
      percUniq <- data.frame(
        GO.ID = id,
        Term = t,
        unique_genes = uniq_gen, 
        total_genes = tot_gen,
        perc_unique = uniq_gen / tot_gen,
        Population = g, 
        Contrast = "Treatment")
      
      percSharedUniq_go_trt <- rbind(percSharedUniq_go_trt, percUniq)
  }
}

# Phenotype
for (t in unique(shared_goTerms_phe_LFC$Grouped_Term)){
  
  tmp <- subset(shared_goTerms_phe_LFC, Grouped_Term == t)
  id <- unique(tmp$Grouped_GO.ID)
    
    for(g in genetic_group){
      
      uniq <- attr(venn(list(
        Guiana = as.character(subset(tmp, Population == "Guiana")$Gene),
        Iquitos = as.character(subset(tmp, Population == "Iquitos")$Gene),
        Marañón = as.character(subset(tmp, Population == "Marañón")$Gene),
        Nanay = as.character(subset(tmp, Population == "Nanay")$Gene)), 
        show.plot = FALSE),"intersections")
      uniq_gen <- as.numeric(length(uniq[[g]]))
      tot_gen <- nrow(subset(tmp, Population == g))
      
      percUniq <- data.frame(
        GO.ID = id,
        Term = t,
        unique_genes = uniq_gen, 
        total_genes = tot_gen,
        perc_unique = uniq_gen / tot_gen,
        Population = g, 
        Contrast = "Phenotype")
      
      percSharedUniq_go_phe <- rbind(percSharedUniq_go_phe, percUniq)
  }
}

# pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_avgPercUniq_sharedGOTerms.pdf")
# rbind(
#     percSharedUniq_go_trt %>% 
#       na.omit() %>%
#       group_by(Population) %>%
#       add_tally() %>%
#       summarize(mean_uniq = mean(perc_unique),
#                 sd_uniq = sd(perc_unique),
#                 se_uniq = sd_uniq / sqrt(n)) %>%
#       distinct() %>%
#       add_column(Contrast = rep("Treatment", nrow(.))),
#     
#     percSharedUniq_go_phe %>% 
#       na.omit() %>%
#       group_by(Population) %>%
#       add_tally() %>%
#       summarize(mean_uniq = mean(perc_unique),
#                 sd_uniq = sd(perc_unique),
#                 se_uniq = sd_uniq / sqrt(n)) %>%
#       distinct() %>%
#       add_column(Contrast = rep("Phenotype", nrow(.)))
#   ) %>%
#   arrange(desc(Contrast)) %>%
#   mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>% 
#   ggplot(., aes(x = Population, y = mean_uniq, color = Population)) +
#   geom_point(size = 4, shape = 15) +
#   scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
#   geom_errorbar(aes(ymin = mean_uniq - se_uniq, ymax = mean_uniq + se_uniq), lwd = 2) +
#   theme_minimal_grid() +
#   labs(x = NULL, y = "% Unique Genes in each GO Term (Average)", color = "Genetic Group") +
#   facet_wrap(~Contrast, ncol = 1, nrow = 2) +
#   theme(
#     axis.text.x = element_text(face = "bold", angle = 45, vjust = 0.5), 
#     axis.text.y = element_text(face = "bold"),
#     strip.background = element_rect(color = "lightgrey")
#   )


# pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_avgPercUniq_sharedGOTerms.pdf", width = 12)
# 
# p1 <- 
#   rbind(
#     percSharedUniq_go_trt %>%
#       add_column(., Contrast = rep("Treatment", nrow(.))),
#     percSharedUniq_go_phe %>%
#       add_column(., Contrast = rep("Phenotype", nrow(.)))
#   
#   ) %>%
#   subset(., Contrast == "Treatment") %>%
#   arrange(desc(Contrast)) %>%
#   mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>% 
#   ggplot(., aes(x = Population, y = perc_unique, color = Population)) +
#   geom_jitter(size = 6, width = 0.15) +
#   stat_summary(fun = mean, color = "blue", shape = 18, size = 2) +
#   scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
#   theme_minimal_grid() +
#   labs(x = NULL, y = "% Unique Genes in each GO Term", color = "Genetic Group") +
#   facet_wrap(~Contrast, ncol = 1, nrow = 2, ) +
#   theme(
#     axis.text.x = element_text(size = 25, face = "bold", angle = 45, vjust = 0.5), 
#     axis.text.y = element_text(size = 25, face = "bold"),
#     axis.title.y = element_text(size = 25, face = "bold"),
#     strip.background = element_rect(color = "lightgrey"),
#     legend.position = "none",
#     strip.text = element_text(size = 30)
#   )
# p1
# 
# p2 <- 
#     rbind(
#     percSharedUniq_go_trt %>%
#       add_column(., Contrast = rep("Treatment", nrow(.))),
#     percSharedUniq_go_phe %>%
#       add_column(., Contrast = rep("Phenotype", nrow(.)))
#   
#   ) %>%
#   subset(., Contrast == "Phenotype") %>%
#   arrange(desc(Contrast)) %>%
#   mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>% 
#   ggplot(., aes(x = Population, y = perc_unique, color = Population)) +
#   geom_jitter(size = 6, width = 0.15) +
#   stat_summary(fun = mean, color = "blue", shape = 18, size = 2) +
#   scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
#   theme_minimal_grid() +
#   labs(x = NULL, y = "% Unique Genes in each GO Term", color = "Genetic Group") +
#   facet_wrap(~Contrast, ncol = 1, nrow = 2, ) +
#   theme(
#     axis.text.x = element_text(size = 25, face = "bold", angle = 45, vjust = 0.5), 
#     axis.text.y = element_text(size = 25, face = "bold"),
#     axis.title.y = element_text(size = 25, face = "bold"),
#     strip.background = element_rect(color = "lightgrey"),
#     legend.text = element_text(size = 25, face = "bold"),
#     legend.title = element_text(size = 25, face = "bold"),
#     legend.position = "right",
#     strip.text = element_text(size = 30)
#   )
# p2
# 
# dev.off()



# Color each go term
df <- rbind(percSharedUniq_go_trt, percSharedUniq_go_phe) %>% 
  mutate(Comb_Term = paste0(GO.ID, ": ", Term))
  

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_avgPercUniq_sharedGOTerms.pdf", height = 14, width = 10)

  ggplot(df, aes(x = Population, y = perc_unique, color = Comb_Term)) +
    geom_jitter(size = 5, width = 0.15) +
    stat_summary(fun = mean, color = "black", shape = 5, size = 2) +
    #scale_color_manual(values = wes_palette("Darjeeling1", type = "discrete")) + 
    #scale_colour_viridis_d(option = "plasma") +
    scale_color_manual(
      values = c(
        "GO:0002238: response to molecule of fungal origin" = '#77AADD', 
        "GO:0009625: response to insect"  = '#EE8866', 
        "GO:0009682: induced systemic resistance" = '#EEDD88', 
        "GO:0009739: response to gibberellin" = '#FFAABB', 
        "GO:0009753: response to jasmonic acid" = '#99DDFF', 
        "GO:0009800: cinnamic acid biosynthetic process" = '#44BB99',
        "GO:0009809: lignin biosynthetic process" = '#BBCC33',
        "GO:0009834: plant-type secondary cell wall biogenesi..." = '#AAAA00',
        "GO:0046686: response to cadmium ion" = '#1a1919')
      ) +
    theme_minimal_grid() +
    labs(x = NULL, y = "% Unique Genes in each GO Term", color = "GO Term") +
    facet_wrap(~Contrast, ncol = 1, nrow = 2, ) +
    theme(
      axis.text.x = element_text(size = 20, angle = 45, vjust = 0.5), 
      axis.text.y = element_text(size = 20),
      axis.title.y = element_text(size = 20),
      strip.background = element_rect(color = "lightgrey"),
      legend.position = "bottom",
      legend.direction = "vertical",
      legend.text = element_text(size = 20),
      legend.title = element_text(size = 20, face = "bold"),
      strip.text = element_text(size = 25),
      panel.spacing = unit(4, "line")
  )
dev.off()



```

#### PLOT -- LFC and expression for highly shared GO terms 
```{r}
# Expression
shared_goTerms_exprTrt <- data.frame()
shared_goTerms_exprPhe <- data.frame()
for (t in unique(shared_goTerms_trt_LFC$GO.ID)){
  tmp <- 
    shared_goTerms_trt_LFC %>%
    subset(., GO.ID == t) %>%
    dplyr::select(., Gene) %>%
    unique() %>%
    inner_join(., norm_expr_trt, by = "Gene") %>% 
    pivot_longer(cols = c(Full, Guiana, Iquitos, Maranon, Nanay), 
                 names_to = "Population", values_to = "Expression") %>%
    add_column(Term = rep(t, nrow(.)))
  
  shared_goTerms_exprTrt <- rbind(shared_goTerms_exprTrt, tmp)

}
shared_goTerms_exprTrt <- filter(shared_goTerms_exprTrt, Population != "Full")

for (t in unique(shared_goTerms_phe_LFC$GO.ID)){
  tmp <- 
    shared_goTerms_phe_LFC %>%
    subset(., GO.ID == t) %>%
    dplyr::select(., Gene) %>%
    unique() %>%
    inner_join(., norm_expr_phe, by = "Gene") %>% 
    pivot_longer(cols = c(Full, Guiana, Iquitos, Maranon, Nanay), 
                 names_to = "Population", values_to = "Expression") %>%
    add_column(Term = rep(t, nrow(.)))
  
  shared_goTerms_exprPhe <- rbind(shared_goTerms_exprPhe, tmp)

}
shared_goTerms_exprPhe <- filter(shared_goTerms_exprPhe, Population != "Full")

# LFC 
shared_goTerms_lfcTrt <- data.frame()
shared_goTerms_lfcPhe <- data.frame()
for (t in unique(shared_goTerms_trt_LFC$GO.ID)){
  tmp <- 
    shared_goTerms_trt_LFC %>%
    subset(., GO.ID == t) %>%
    dplyr::select(., Gene) %>%
    unique() %>%
    inner_join(., TmainLFC_all_pop, by = "Gene") %>% 
    # pivot_longer(cols = c(Guiana, Iquitos, Maranon, Nanay), 
    #              names_to = "Population", values_to = "Expression") %>%
    add_column(Term = rep(t, nrow(.)))
  
  shared_goTerms_lfcTrt <- rbind(shared_goTerms_lfcTrt, tmp)

}


for (t in unique(shared_goTerms_phe_LFC$GO.ID)){
  tmp <- 
    shared_goTerms_phe_LFC %>%
    subset(., GO.ID == t) %>%
    dplyr::select(., Gene) %>%
    unique() %>%
    inner_join(., PmainLFC_all_pop, by = "Gene") %>% 
    # pivot_longer(cols = c(Guiana, Iquitos, Maranon, Nanay), 
    #              names_to = "Population", values_to = "Expression") %>%
    add_column(Term = rep(t, nrow(.)))
  
  shared_goTerms_lfcPhe <- rbind(shared_goTerms_lfcPhe, tmp)

}


trt_expr_mods <- data.frame()
for (i in unique(shared_goTerms_exprTrt$Term)){
  tmp <- subset(shared_goTerms_exprTrt, Term == i & Population != "Full")
  mod <- aov(Expression ~ Population, tmp)
  mod <- tidy(mod) %>%
    add_column(Term = rep(i, nrow(.)), 
               Comparison = rep("Expr", nrow(.)), 
               Effect = rep("Treatment", nrow(.))) %>%
    filter(term == "Population")
  trt_expr_mods <- rbind(trt_expr_mods, mod)
}

trt_lfc_mods <- data.frame()
for (i in unique(shared_goTerms_lfcTrt$Term)){
  tmp <- subset(shared_goTerms_lfcTrt, Term == i & Population != "Full")
  mod <- aov(log2FoldChange ~ Population, tmp)
  mod <- tidy(mod) %>%
    add_column(Term = rep(i, nrow(.)), 
               Comparison = rep("LFC", nrow(.)), 
               Effect = rep("Treatment", nrow(.))) %>%
    filter(term == "Population")
  trt_lfc_mods <- rbind(trt_lfc_mods, mod)
}

phe_expr_mods <- data.frame()
for (i in unique(shared_goTerms_exprPhe$Term)){
  tmp <- subset(shared_goTerms_exprPhe, Term == i & Population != "Full")
  mod <- aov(Expression ~ Population, tmp) 
  mod <- tidy(mod) %>%
    add_column(Term = rep(i, nrow(.)), 
               Comparison = rep("Expr", nrow(.)), 
               Effect = rep("Phenotype", nrow(.))) %>%
    filter(term == "Population")
  phe_expr_mods <- rbind(phe_expr_mods, mod)
}

phe_lfc_mods <- data.frame()
for (i in unique(shared_goTerms_lfcPhe$Term)){
  tmp <- subset(shared_goTerms_lfcPhe, Term == i & Population != "Full")
  mod <- aov(log2FoldChange ~ Population, tmp)
  mod <- tidy(mod) %>%
    add_column(Term = rep(i, nrow(.)), 
               Comparison = rep("LFC", nrow(.)), 
               Effect = rep("Phenotype", nrow(.))) %>%
    filter(term == "Population")
  phe_lfc_mods <- rbind(phe_lfc_mods, mod)
}

allMods <- rbind(trt_expr_mods, trt_lfc_mods, 
           phe_expr_mods, phe_lfc_mods)
allMods$padj <- formatC(p.adjust(allMods$p.value, method = "fdr"), format = 'e', digits = 2)


pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_exprLFC_sharedGOTerms.pdf", width = 14, height = 14)
p1 <-
  shared_goTerms_exprTrt %>% 
  ggplot(aes(x = Population, y = Expression, color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "Normalized Expression (VST)", title = "TREATMENT") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45, hjust = 1)
    ) + 
  #scale_color_manual(values = wes_palette("Darjeeling1", type = "discrete")) +
  scale_color_manual(values = c("GO:0002238" = '#77AADD', "GO:0009625" = '#EE8866', 
                                "GO:0009682" = '#EEDD88', "GO:0009739" = '#FFAABB', 
                                "GO:0009753" = '#99DDFF', "GO:0009800" = '#44BB99', 
                                "GO:0009809" = '#BBCC33', "GO:0009834" = '#AAAA00',
                                "GO:0046686" = '#1a1919')) +
  facet_wrap(~Term) + 
  geom_text(data = subset(allMods, Comparison == "Expr" & Effect == "Treatment"), 
            aes(x = 2, y = 12.9, label = paste0("p = ", padj)), color = "black", size = 5)

p2 <-
    shared_goTerms_exprPhe %>% 
    ggplot(aes(x = Population, y = Expression, color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "Normalized Expression (VST)", title = "PHENOTYPE") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45, hjust = 1)
    ) + 
  #scale_color_manual(values = wes_palette("Darjeeling1", type = "discrete")) +
    scale_color_manual(values = c("GO:0002238" = '#77AADD', "GO:0009625" = '#EE8866', 
                                "GO:0009682" = '#EEDD88', "GO:0009739" = '#FFAABB', 
                                "GO:0009753" = '#99DDFF', "GO:0009800" = '#44BB99', 
                                "GO:0009809" = '#BBCC33', "GO:0009834" = '#AAAA00',
                                "GO:0046686" = '#1a1919')) +
    facet_wrap(~Term) +
    geom_text(data = subset(allMods, Comparison == "Expr" & Effect == "Phenotype"), 
            aes(x = 1.5, y = 12.5, label = paste0("p = ", padj)), color = "black", size = 5)

# LFC
p3 <-
    shared_goTerms_lfcTrt %>%
    ggplot(aes(x = Population, y = log2FoldChange, color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "log2[Fold Change]", title = "TREATMENT") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45, hjust = 1)
    ) +
    scale_color_manual(values = c("GO:0002238" = '#77AADD', "GO:0009625" = '#EE8866', 
                                "GO:0009682" = '#EEDD88', "GO:0009739" = '#FFAABB', 
                                "GO:0009753" = '#99DDFF', "GO:0009800" = '#44BB99', 
                                "GO:0009809" = '#BBCC33', "GO:0009834" = '#AAAA00',
                                "GO:0046686" = '#1a1919')) +
  #scale_color_manual(values = wes_palette("Darjeeling1", type = "discrete")) +
  # geom_text(aes(x = 5, y = (min(log2FoldChange) + 1),
  #               label = mean(log2FoldChange)), 
  #           color = "black") +   
  # geom_text(aes(x = 5, y = min(log2FoldChange), 
  #               label = sd(log2FoldChange)), 
  #           color = "black") +
    facet_wrap(~Term) + 
    geom_text(data = subset(allMods, Comparison == "LFC" & Effect == "Treatment"), 
              aes(x = 2, y = 3.8, label = paste0("p = ", padj)), color = "black", size = 5)

p4 <-
    shared_goTerms_lfcPhe %>%
    ggplot(aes(x = Population, y = log2FoldChange,color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "log2[Fold Change]", title = "PHENOTYPE") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45, hjust = 1)
    ) +
    scale_color_manual(values = c("GO:0002238" = '#77AADD', "GO:0009625" = '#EE8866', 
                                "GO:0009682" = '#EEDD88', "GO:0009739" = '#FFAABB', 
                                "GO:0009753" = '#99DDFF', "GO:0009800" = '#44BB99', 
                                "GO:0009809" = '#BBCC33', "GO:0009834" = '#AAAA00',
                                "GO:0046686" = '#1a1919')) +
  #scale_color_manual(values = wes_palette("Darjeeling1", type = "discrete")) +
  # geom_text(aes(x = 5, y = (min(log2FoldChange) + 1),
  #               label = mean(log2FoldChange)), 
  #           color = "black") +   
  # geom_text(aes(x = 5, y = min(log2FoldChange), 
  #               label = sd(log2FoldChange)), 
  #           color = "black") + 
    facet_wrap(~Term) + 
    geom_text(data = subset(allMods, Comparison == "LFC" & Effect == "Phenotype"), 
              aes(x = 1.5, y = 3.6, label = paste0("p = ", padj)), color = "black", size = 5)

plot_grid(p3, p1, p4, p2, ncol = 2, 
          nrow = 2)

dev.off()

```

#### PLOT -- LFC correlations for highly shared GO terms
```{r}
shared_goTerms_lfcTrt_corr <- 
  shared_goTerms_lfcTrt %>% 
  dplyr::select(Gene, log2FoldChange, Population) %>%
  distinct() %>%
  pivot_wider(names_from = Population, values_from = log2FoldChange) %>%
  column_to_rownames(var = "Gene") 

shared_goTerms_lfcPhe_corr <- 
  shared_goTerms_lfcPhe %>% 
  dplyr::select(Gene, log2FoldChange, Population) %>%
  distinct() %>%
  pivot_wider(names_from = Population, values_from = log2FoldChange) %>%
  column_to_rownames(var = "Gene") 

shared_goTerms_exprTrt_corr <- 
  shared_goTerms_exprTrt %>%
  dplyr::select(Gene, Expression, Population) %>%
  distinct() %>%
  pivot_wider(names_from = Population, values_from = Expression) %>%
  column_to_rownames(var = "Gene") 

shared_goTerms_exprPhe_corr <- 
  shared_goTerms_exprPhe %>% 
  dplyr::select(Gene, Expression, Population) %>%
  distinct() %>%
  pivot_wider(names_from = Population, values_from = Expression) %>%
  column_to_rownames(var = "Gene") 


shared_go_trt_pmat <- cor.mtest(shared_goTerms_lfcTrt_corr, alternative = "two.sided", method = "spearman")
shared_go_trt_pmat$p[lower.tri(shared_go_trt_pmat$p)] <- 1

shared_go_phe_pmat <- cor.mtest(shared_goTerms_lfcPhe_corr, alternative = "two.sided", method = "spearman")
shared_go_phe_pmat$p[lower.tri(shared_go_phe_pmat$p)] <- 1

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig5_log2FC_corrMatrix_sharedGOTerms.pdf", height = 25, width = 30)
par(mfrow=c(2,2))

shared_go_trt_pmat <- cor.mtest(shared_goTerms_lfcTrt_corr, alternative = "two.sided", method = "spearman")
shared_go_trt_pmat$p[lower.tri(shared_go_trt_pmat$p)] <- 1

shared_go_phe_pmat <- cor.mtest(shared_goTerms_lfcPhe_corr, alternative = "two.sided", method = "spearman")
shared_go_phe_pmat$p[lower.tri(shared_go_phe_pmat$p)] <- 1

# LFC correlation matrices for SHARED GO terms -- Treatment
corrplot.mixed(cor(shared_goTerms_lfcTrt_corr, method = "spearman"), p.mat = shared_go_trt_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

# LFC correlation matrices for SHARED GO terms -- Phenotype
corrplot.mixed(cor(shared_goTerms_lfcPhe_corr, method = "spearman"), p.mat = shared_go_phe_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

shared_go_trt_pmat <- cor.mtest(shared_goTerms_exprTrt_corr, alternative = "two.sided", method = "spearman")
shared_go_trt_pmat$p[lower.tri(shared_go_trt_pmat$p)] <- 1

shared_go_phe_pmat <- cor.mtest(shared_goTerms_exprPhe_corr, alternative = "two.sided", method = "spearman")
shared_go_phe_pmat$p[lower.tri(shared_go_phe_pmat$p)] <- 1

# Expression correlation matrices for SHARED GO terms -- Treatment
corrplot.mixed(cor(shared_goTerms_exprTrt_corr, method = "spearman"), p.mat = shared_go_trt_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

# Expression correlation matrices for SHARED GO terms -- Phenotype
corrplot.mixed(cor(shared_goTerms_exprPhe_corr, method = "spearman"), p.mat = shared_go_phe_pmat$p, insig = 'label_sig', sig.level = c(0.001, 0.01, 0.05), pch.cex = 3,  tl.col = c("steelblue", "firebrick", "darkgreen", "darkorange"), upper = "ellipse",  order = "original", number.cex = 3, number.font = 3, lower.col = "black", tl.srt = 45, tl.cex = 3, cl.cex = 2)

dev.off()

```

#### PLOT -- Expression and LFC for highly shared GO terms -- lignin biosynthesis
```{r}
pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_exprLFC_sharedGOTerms_ligninBiosynth.pdf", width = 7, height = 7)

p1 <-
  shared_goTerms_exprTrt %>% 
  subset(Term == "GO:0009809") %>%
    ggplot(aes(x = Population, y = Expression, color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "Normalized Expression (VST)", title = "TREATMENT") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45)
    ) + 
  scale_color_manual(values = '#BBCC33') +
  facet_wrap(~Term)

p2 <-
  shared_goTerms_exprPhe %>% 
  subset(Term == "GO:0009800") %>%
    ggplot(aes(x = Population, y = Expression, color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "Normalized Expression (VST)", title = "PHENOTYPE") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45)
    ) + 
  scale_color_manual(values = '#44BB99') +
  facet_wrap(~Term)

# LFC
p3 <-
  shared_goTerms_lfcTrt %>% 
    subset(Term == "GO:0009809") %>%
    ggplot(aes(x = Population, y = log2FoldChange, color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "log2[Fold Change]", title = "TREATMENT") + 
    theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45)
    ) +
    scale_color_manual(values = '#BBCC33') +
  # geom_text(aes(x = 5, y = (min(log2FoldChange) + 1),
  #               label = mean(log2FoldChange)), 
  #           color = "black") +   
  # geom_text(aes(x = 5, y = min(log2FoldChange), 
  #               label = sd(log2FoldChange)), 
  #           color = "black") +
  facet_wrap(~Term)

p4 <-
  shared_goTerms_lfcPhe %>%
  subset(Term == "GO:0009800") %>%
    ggplot(aes(x = Population,
               y = log2FoldChange,
               color = Term)) +
    geom_boxplot() + 
    geom_jitter(width = 0.25, alpha = 0.5) + 
    #scale_color_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
    labs(x = NULL, y = "log2[Fold Change]", title = "PHENOTYPE") + 
   theme_minimal_grid() + 
    theme(
      legend.position = "none",
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45)
    ) +
    scale_color_manual(values = '#44BB99') +
  # geom_text(aes(x = 5, y = (min(log2FoldChange) + 1),
  #               label = mean(log2FoldChange)), 
  #           color = "black") +   
  # geom_text(aes(x = 5, y = min(log2FoldChange), 
  #               label = sd(log2FoldChange)), 
  #           color = "black") + 
  facet_wrap(~Term)


plot_grid(p3, p1, p4, p2, ncol = 2, 
          nrow = 2)
dev.off()
```

#### PLOT -- Interesting genes shared GO terms
```{r}
# What genes show up in BOTH treatment and phenotype, AND are found in >= 3 populations?
# Treatment
TmainLFC_top1000_sharedPops <- 
  TmainLFC_all_pop_lfc1000 %>% 
  group_by(Gene) %>% 
  add_tally() %>%
  subset(n >= 3) %>% 
  group_by(Gene) %>% 
  summarise(mLFC = mean(log2FoldChange),
            mSD = sd(log2FoldChange), 
            CV = (mSD / abs(mLFC)) * 100,
            n = n()) %>% 
  mutate(CV = round(CV, digits = 2)) %>%
  inner_join(annot, by = "Gene") %>% 
  mutate(abs = abs(mLFC)) %>%
  setNames(nm = c("Gene", paste0(colnames(.)[-1], "_treatment")))
  
# Phenotype
PmainLFC_top1000_sharedPops <- 
  PmainLFC_all_pop_lfc1000 %>% 
  group_by(Gene) %>% 
  add_tally() %>% 
  subset(n >= 3) %>% 
  group_by(Gene) %>% 
  summarise(mLFC = mean(log2FoldChange),
            mSD = sd(log2FoldChange), 
            CV = (mSD / abs(mLFC)) * 100,
            n = n()) %>% 
  mutate(CV = round(CV, digits = 2)) %>%
  inner_join(annot, by = "Gene") %>% 
  mutate(abs = abs(mLFC)) %>% 
  setNames(nm = c("Gene", paste0(colnames(.)[-1], "_phenotype")))

# Overlap
TmainLFC_PmainLFC_sharedPops <- inner_join(TmainLFC_top1000_sharedPops,
                                           PmainLFC_top1000_sharedPops, by = "Gene")

# Write out table
write.table(TmainLFC_PmainLFC_sharedPops, 
            "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/TableS3_treatPheno_topLFC.tsv", sep = '\t', quote = FALSE, row.names = FALSE)




# Interesting genes from conserved GO terms--TREATMENT
shared_genes_trt_consGOTerms <- data.frame()
for (t in unique(shared_goTerms_trt_LFC$Grouped_Term)){
  
  tmp <- subset(shared_goTerms_trt_LFC, Grouped_Term == t)
  cat(paste0(t, "\n"))
  uniq <- list(
    Guiana = as.character(subset(tmp, Population == "Guiana")$Gene),
    Iquitos = as.character(subset(tmp, Population == "Iquitos")$Gene),
    Marañón = as.character(subset(tmp, Population == "Marañón")$Gene),
    Nanay = as.character(subset(tmp, Population == "Nanay")$Gene))
  df <- list_to_matrix(uniq)
  df <- df %>% 
    data.frame() %>%
    mutate(population_sum = rowSums(.)) %>%
    subset(population_sum > 1) %>% 
    rownames_to_column(var = "Gene") %>%
    mutate(Term = rep(t, nrow(.))) %>%
    inner_join(annot, by = "Gene") %>%
    dplyr::select(c(Gene, Term, population_sum, Desc))
  
  shared_genes_trt_consGOTerms <- 
    rbind(df, shared_genes_trt_consGOTerms)
}


# Interesting genes from conserved GO terms--PHENOTYPE
shared_genes_phe_consGOTerms <- data.frame()
for (t in unique(shared_goTerms_phe_LFC$Grouped_Term)){
  
  tmp <- subset(shared_goTerms_phe_LFC, Grouped_Term == t)
  cat(paste0(t, "\n"))
  uniq <- list(
    Guiana = as.character(subset(tmp, Population == "Guiana")$Gene),
    Iquitos = as.character(subset(tmp, Population == "Iquitos")$Gene),
    Marañón = as.character(subset(tmp, Population == "Marañón")$Gene),
    Nanay = as.character(subset(tmp, Population == "Nanay")$Gene))
  df <- list_to_matrix(uniq)
  df <- df %>% 
    data.frame() %>%
    mutate(population_sum = rowSums(.)) %>%
    subset(population_sum > 1) %>% 
    rownames_to_column(var = "Gene") %>%
    mutate(Term = rep(t, nrow(.))) %>%
    inner_join(annot, by = "Gene") %>%
    dplyr::select(c(Gene, Term, population_sum, Desc))
  
  shared_genes_phe_consGOTerms <- 
    rbind(df, shared_genes_phe_consGOTerms)
  
}

shared_genes_both_consGOTerms <- 
  rbind(shared_genes_trt_consGOTerms, shared_genes_phe_consGOTerms) %>% 
  mutate(contrast = ifelse(Gene %in% shared_genes_trt_consGOTerms$Gene, "Treatment", "Phenotype")) %>%
  mutate(contrast = ifelse(Gene %in% shared_genes_trt_consGOTerms$Gene & Gene %in% shared_genes_phe_consGOTerms$Gene,
                           "Both", 
                           contrast))
 
# Interesting, non-lignin genes
# WRKYs, MYBs, Other -- shared among 4 genetic groups AND Treatment + Phenotype for our cons GO terms:	
int_genes <- 
  c("SCA-6_Chr2v1_07442.1", "SCA-6_Chr3v1_09222.1",
    "SCA-6_Chr3v1_10161.1",	"SCA-6_Chr1v1_03120.1",
    "SCA-6_Chr4v1_12269.1", "SCA-6_Chr6v1_16922.1",
    "SCA-6_Chr9v1_23321.1", "SCA-6_Chr7v1_20558.1",
    "SCA-6_Chr7v1_19998.1")
int_genes <- norm_expr_long[norm_expr_long$SCA6_Gene %in% int_genes,] %>%
  setNames(nm = c("Gene", "sample", "expression")) %>%
  inner_join(shared_genes_both_consGOTerms, by = "Gene")

# Interesting, non-lignin genes
pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/non_lignin_geneExpr.pdf", 
    height = 12, width = 12)
colData %>% 
  rownames_to_column(var = "sample") %>% 
  mutate(accession = gsub("_[0-9].*", "", sample)) %>%
  inner_join(., int_genes, by = "sample") %>% 
  mutate(gene_name = gsub("_.*RecName: Full=.*", "", Desc)) %>% 
  dplyr::select(c(Treatment, expression, Phenotype, gene_name, contrast, Gen_Group)) %>%
  unique() %>%
  ggplot(aes(x = Treatment, 
               y = expression, 
               fill = Treatment, 
               shape = Phenotype, 
              group = Treatment)) +
  #geom_violin(width = 0.5, show.legend = FALSE) + 
    geom_boxplot(width = 0.2, show.legend = FALSE) +
    geom_jitter(width = 0.2, size = 2, alpha = 0.75) + 
    #scale_fill_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
    scale_fill_manual(values = c("#3B9AB2","#E1AF00")) +
    scale_x_discrete(labels=c("control" = "V8 Control", "treatment" = "P. palmivora\n Treatment")) +
    scale_shape_manual(values = c(21, 24)) +
    scale_color_manual(values = "black") +
    #stat_summary(fun = mean, color = "black", shape = 5, show.legend = FALSE) +
    facet_grid(gene_name + contrast ~ Gen_Group, scales = "free_y") + 
    theme_minimal_grid(font_size = 10) + 
    labs(x = NULL, y = "Normalized Expression (VST)", , fill = NULL, shape = NULL) +
    theme(
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45, hjust = 1),
      axis.title.y = element_text(size = 10),
      title = element_text(size = 10), 
      legend.justification = "center",
      legend.position = "bottom",
      legend.text = element_text(size = 10)
  ) +
    guides(fill = guide_legend(override.aes = list(shape = 21, size=5)),
           shape = guide_legend(override.aes = list(fill = "black", size=5)))
dev.off()


# Interesting, lignin-related genes
int_genes <- c("SCA-6_Chr5v1_15983.1", "SCA-6_Chr5v1_15983.1",
               "SCA-6_Chr9v1_22876.1","SCA-6_Chr6v1_17513.1")

int_genes <- norm_expr_long[norm_expr_long$SCA6_Gene %in% int_genes,] %>% 
  setNames(nm = c("Gene", "sample", "expression")) %>%
  inner_join(shared_genes_both_consGOTerms, by = "Gene")


# Interesting, lignin-related genes
pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/lignin_geneExpr.pdf", 
    height = 12, width = 12)
colData %>% 
  rownames_to_column(var = "sample") %>% 
  mutate(accession = gsub("_[0-9].*", "", sample)) %>%
  inner_join(., int_genes, by = "sample") %>% 
  mutate(gene_name = gsub("_.*RecName: Full=.*", "", Desc)) %>% 
  dplyr::select(c(Treatment, expression, Phenotype, gene_name, contrast, Gen_Group)) %>%
  unique() %>%
  ggplot(aes(x = Treatment, 
               y = expression, 
               fill = Treatment, 
               shape = Phenotype, 
              group = Treatment)) +
  #geom_violin(width = 0.5, show.legend = FALSE) + 
    geom_boxplot(width = 0.2, show.legend = FALSE) +
    geom_jitter(width = 0.2, size = 2, alpha = 0.75) + 
    #scale_fill_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
    scale_fill_manual(values = c("#3B9AB2","#E1AF00")) +
    scale_x_discrete(labels=c("control" = "V8 Control", "treatment" = "P. palmivora\n Treatment")) +
    scale_shape_manual(values = c(21, 24)) +
    scale_color_manual(values = "black") +
    #stat_summary(fun = mean, color = "black", shape = 5, show.legend = FALSE) +
    facet_grid(gene_name + contrast ~ Gen_Group, scales = "free_y") + 
    theme_minimal_grid(font_size = 10) + 
    labs(x = NULL, y = "Normalized Expression (VST)", fill = NULL, shape = NULL) +
    theme(
      strip.background = element_rect(color = "lightgrey"),
      axis.text.x = element_text(angle = 45, hjust = 1),
      axis.title.y = element_text(size = 10),
      title = element_text(size = 10), 
      legend.justification = "center",
      legend.position = "bottom",
      legend.text = element_text(size = 10)
  ) +
    guides(fill = guide_legend(override.aes = list(shape = 21, size=5)),
           shape = guide_legend(override.aes = list(fill = "black", size=5)))
dev.off()



# x <- 
#   shared_goTerms_lfcTrt %>%
#   dplyr::select(Gene, Population, log2FoldChange) %>% 
#   inner_join(annot, by = "Gene")  %>% 
#   group_by(Gene) %>% 
#   add_tally()
#   .[!(duplicated(.)|duplicated(., fromLast=TRUE)),]
# 
#   mutate(abbrev_annot = case_when(
#     grepl("MYB", Desc) ~ "MYB TF", 
#     grepl("Laccase", Desc) ~ "Laccase", 
#     grepl("WRKY", Desc) ~ "WRKY TF",
#     
#   )) %>% 
#   filter(abbrev_annot == "MYB TF" | abbrev_annot == "Laccase" | abbrev_annot == "WRKY TF") %>%
#   group_by(Population, abbrev_annot) %>% 
#   summarise(max(log2FoldChange))
#   
# # Create long-form expression matrix
# norm_expr_long <- 
#   norm_expr %>% 
#   data.frame(.) %>%
#   rownames_to_column(var = "SCA6_Gene") %>% 
#   pivot_longer(cols = -SCA6_Gene, 
#                names_to = "sample", 
#                values_to = "expression")
# 
# 

```








## PBS Analyses 
### Pre-processing PBS results, read in data
```{r}
# Load in PBS dataframes as list
pbs_list <- list.files(path = "~/GradSchool/Research/Projects/NSF_PlantGenome/PBS_Selection_Windows_Analysis/PBS_results_Feb2021/", pattern = "*.txt", full.names = TRUE)

pbs_df <- lapply(pbs_list, function(x){
  read.table(x, header = TRUE)
})

# Substitute the path to just leave the genetic groups 
pbs_list <- gsub("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/PBS_Selection_Windows_Analysis/PBS_results_Feb2021//", "", pbs_list) %>% 
  sub(".pbs.txt", "", .)

# Name the dataframes
names(pbs_df) <- pbs_list

# df <- read.table("~/Downloads/PBS_results/Iquitos.pbs.txt", header = TRUE)
for (i in 1:length(pbs_df)){
  
  tmp_df <- data.frame(pbs_df[[i]])
  colnames(tmp_df)[1] <- "Gene"
  tmp_df <- 
    tmp_df %>%
    # inner_join(., gene_pos, by = "Gene")%>%
    # add_column(prime5_reg_pbs = ifelse(.$Strand == "-", .$down_pbs, .$up_pbs)) %>%
    # add_column(prime3_reg_pbs = ifelse(.$Strand == "-", .$up_pbs, .$down_pbs)) %>% 
    melt(., id.var = "Gene", ) %>%
    mutate(., value = as.numeric(value)) %>%
    mutate(., variable = as.character(variable)) %>%
    mutate(., Gene = as.character(Gene)) %>%
    .[is.finite(.$value),] %>%
    subset(., subset = variable == "coding_pbs" | variable == "up_pbs" | variable == "down_pbs")
  pbs_df[[i]] <- tmp_df
  }


# df <- na.omit(df) %>%
#   subset(., coding_pbs != "Inf") %>%
#   subset(., coding_pbs != "-Inf") %>%
#   melt(., id.var = "id") %>%
#   subset(., subset = variable == "coding_pbs" | variable == "up_pbs" | variable == "down_pbs")

n <- 0.01
list_plots <- list()
for (z in 1:length(pbs_df)){
  df <- data.frame(pbs_df[[z]])
  vline <- c()
  
  for (i in unique(df$variable)){
    val <-
      df %>%
      subset(., variable == i) %>%
      arrange(., desc(value)) %>%
      .[ceiling(n * nrow(.)),] %>%
      .$value
    vline[[i]] <- val
  }
  print(mean(vline))
  
  p <-
  ggplot(df) +
    geom_density( aes(x=value, fill = variable), alpha=0.4) + 
    theme_minimal_grid() +
    geom_vline(aes(xintercept = mean(vline)), col = "red", linetype = "dashed") + 
    labs(x = "log10[# SNP]", y = "Frequency", title = toupper(pbs_list[[z]]))
  
  list_plots[[z]] <- p
}

# pdf("~/Desktop/diff_exp_figs/pbs_analyses/pbs_codingNoncoding_densityPBS.pdf", width = 12)
plot <- gridExtra::grid.arrange(grobs = list_plots)
# dev.off()
```

### Take top n% of PBS values
```{r}
# pbs_list("Guiana", "Iquitos", "Maranon", "Nanay")
pbs_df <- bind_rows(pbs_df, .id = "Group") %>%
  melt(., id_vars = "Gene") %>%
  .[,-4] %>%
  setNames(., nm = c("Group", "Gene", "Type", "PBS")) %>%
  mutate(., Group = ifelse(.$Group == "Maranon", "Marañón", .$Group))

# What percent of genes do we want to look at?
n <- 0.01

# Function to extract top 1% of hits
fTop1 <- function(x){
  head(x[order(x$PBS, decreasing = TRUE) , ], n * nrow(x))
}

# Iterate function over each group and type (coding, upstream etc.)
pbs_df_top1 <- 
  pbs_df %>%
  group_by(Group, Type) %>%
  do(fTop1(.)) 

#pbs_df_top1$Group <- ifelse(pbs_df_top1$Group == "Maranon", "Marañón", pbs_df_top1$Group)
upDownPBS <- subset(pbs_df_top1, Type == "up_pbs" | Type == "down_pbs")

# PBS overlapping treatment and phenotype main effects
TMainLFC_sig <- subset(TMainLFC, padj < 0.05)
PMainLFC_sig <- subset(PMainLFC, padj < 0.05)
ggvenn(
  list( TREATMENT = (TMainLFC_sig %>% mutate(., Gene = gsub("\\.1", "", .$Gene)) %>% .$Gene), 
        FIVE_KB_UP_DOWN = upDownPBS$Gene, 
        PHENOTYPE = (PMainLFC_sig %>%  mutate(., Gene = gsub("\\.1", "", .$Gene)) %>% .$Gene)), 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF"),
  stroke_size = 0.5, text_size = 10, set_name_size = 10
  )

# Overlap between top n% genes in PBS and treatment/phenotype main effects
# sig_gene <-TMainLFC_sig %>%
#   mutate(., Gene = gsub("\\.1", "", .$Gene))
# annot_genes <- annot %>%
#   mutate(., Gene = gsub("\\.1", "", .$Gene))
# 
# coding_sig <- inner_join(subset(pbs_df_top1, Type == "coding_pbs"), sig_gene, by = "Gene") %>%
#   inner_join(., annot_genes, by = "Gene")
# up_sig <- inner_join(subset(pbs_df_top1, Type == "up_pbs"), sig_gene, by = "Gene") %>%
#   inner_join(., annot_genes, by = "Gene")
# down_sig <- inner_join(subset(pbs_df_top1, Type == "down_pbs"), sig_gene, by = "Gene") %>%
#   inner_join(., annot_genes, by = "Gene")
```

#### PLOT -- upset plots for overlap of top n% PBS genes
```{r}
# 5Kb upstream PBS
lt <- list(
  Guiana = as.character(subset(pbs_df_top1, Group == "Guiana" & Type == "up_pbs")$Gene),
  Iquitos = as.character(subset(pbs_df_top1, Group == "Iquitos" & Type == "up_pbs")$Gene),
  Marañón = as.character(subset(pbs_df_top1, Group == "Marañón" & Type == "up_pbs")$Gene),
  Nanay = as.character(subset(pbs_df_top1, Group == "Nanay" & Type == "up_pbs")$Gene)
  )
upPBS_df <- list_to_matrix(lt) %>%
  data.frame(.)

#  Coding PBS
lt <- list(
  Guiana = as.character(subset(pbs_df_top1, Group == "Guiana" & Type == "coding_pbs")$Gene),
  Iquitos = as.character(subset(pbs_df_top1, Group == "Iquitos" & Type == "coding_pbs")$Gene),
  Marañón = as.character(subset(pbs_df_top1, Group == "Marañón" & Type == "coding_pbs")$Gene),
  Nanay = as.character(subset(pbs_df_top1, Group == "Nanay" & Type == "coding_pbs")$Gene)
  )
codingPBS_df <- list_to_matrix(lt) %>%
  data.frame(.)

# 5Kb downstream PBS
lt <- list(
  Guiana = as.character(subset(pbs_df_top1, Group == "Guiana" & Type == "down_pbs")$Gene),
  Iquitos = as.character(subset(pbs_df_top1, Group == "Iquitos" & Type == "down_pbs")$Gene),
  Marañón = as.character(subset(pbs_df_top1, Group == "Marañón" & Type == "down_pbs")$Gene),
  Nanay = as.character(subset(pbs_df_top1, Group == "Nanay" & Type == "down_pbs")$Gene)
  )
downPBS_df <- list_to_matrix(lt) %>%
  data.frame(.)


pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig8_overlapPBS.pdf",
    width = 12)

 # Treatment
 upset(upPBS_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
  #list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Top 1% PBS",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))


 

# Treatment
 upset(codingPBS_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
  #list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Top 1% PBS",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))

# Treatment
 upset(downPBS_df, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T)),
  #list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Top 1% PBS",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))

dev.off()


```

#### PLOT -- correlation between pathogen LFC and PBS within each population
```{r}
PmainLFC_combGroups <- rbind(
  add_column(PmainLFC_guiana, Group = rep("Guiana", nrow(PmainLFC_guiana))) %>%
    add_column(., Sig = ifelse(.$Gene %in% PmainLFC_guiana_lfc1000$Gene, "Significant", "Not Significant")),
  
  add_column(PmainLFC_iquitos, Group = rep("Iquitos", nrow(PmainLFC_iquitos))) %>%
     add_column(., Sig = ifelse(.$Gene %in% PmainLFC_iquitos_lfc1000$Gene, "Significant", "Not Significant")),
  
  add_column(PmainLFC_maranon, Group = rep("Marañón", nrow(PmainLFC_maranon))) %>%
     add_column(., Sig = ifelse(.$Gene %in% PmainLFC_maranon_lfc1000$Gene, "Significant", "Not Significant")),

  add_column(PmainLFC_nanay, Group = rep("Nanay", nrow(PmainLFC_nanay))) %>%
     add_column(., Sig = ifelse(.$Gene %in% PmainLFC_nanay_lfc1000$Gene, "Significant", "Not Significant"))
) %>%
  mutate(., Gene = gsub("\\.1", "", Gene))

# 5-prime correlations
lfc_pbs_df <- data.frame()
for (i in unique(PmainLFC_combGroups$Group)){
  tmp1 <- pbs_df %>%
    subset(., Group == i & Type == "up_pbs") 
  
  tmp2 <- PmainLFC_combGroups  %>%
    subset(., Group == i) %>%
    dplyr::select(., -Group)
  
  tmp3 <- inner_join(tmp1, tmp2, by = "Gene")
  
  lfc_pbs_df <- rbind(lfc_pbs_df, tmp3)
  
}

pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/5prime_5KbPBS_top1000_lfcCorr.pdf", height = 12, width = 6)
my.formula = y ~ x
ggplot(lfc_pbs_df, aes(x = PBS, y = log2FoldChange))  +
  geom_point(data = subset(lfc_pbs_df, Sig == "Significant"), aes(color = Group), alpha = 0.5) +
  geom_point(data = subset(lfc_pbs_df, Sig == "Not Significant"), alpha = 0.1, color = "black") +
  #scale_alpha_discrete(range = c(0.1,1)) + 
  theme_minimal_grid() + 
  # scale_color_manual(values = c( c("grey", "steelblue"), c("grey", "firebrick"),
  #                              c("grey", "darkgreen"), c("grey", "darkorange"))) +
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
  geom_smooth(linetype = 3, color = "red") +
  stat_poly_eq(formula = my.formula, 
               aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               parse = TRUE) +
  theme(
    legend.position = "none"
  ) + 
  facet_wrap(~ Group, ncol = 1, nrow = 4, )
dev.off()

# 3-prime correlations
lfc_pbs_df <- data.frame()
for (i in unique(PmainLFC_combGroups$Group)){
  tmp1 <- pbs_df %>%
    subset(., Group == i & Type == "down_pbs") 
  
  tmp2 <- PmainLFC_combGroups  %>%
    subset(., Group == i) %>%
    dplyr::select(., -Group)
  
  tmp3 <- inner_join(tmp1, tmp2, by = "Gene")
  
  lfc_pbs_df <- rbind(lfc_pbs_df, tmp3)
  
}

pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/3prime_5KbPBS_top1000_lfcCorr.pdf", height = 12, width = 6)
my.formula = y ~ x
ggplot(lfc_pbs_df, aes(x = PBS, y = log2FoldChange))  +
  geom_point(data = subset(lfc_pbs_df, Sig == "Significant"), aes(color = Group), alpha = 0.5) +
  geom_point(data = subset(lfc_pbs_df, Sig == "Not Significant"), alpha = 0.1, color = "black") +
  #scale_alpha_discrete(range = c(0.1,1)) + 
  theme_minimal_grid() + 
  # scale_color_manual(values = c( c("grey", "steelblue"), c("grey", "firebrick"),
  #                              c("grey", "darkgreen"), c("grey", "darkorange"))) +
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
  geom_smooth(linetype = 3, color = "red") +
  stat_poly_eq(formula = my.formula, 
               aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               parse = TRUE) +
  theme(
    legend.position = "none"
  ) + 
  facet_wrap(~ Group, ncol = 1, nrow = 4, )
dev.off()

```

#### PLOT -- correlation between pathogen LFC and PBS within each population -- significant genes
```{r}
PmainLFC_combGroups <- rbind(
  add_column(PmainLFC_guiana, Group = rep("Guiana", nrow(PmainLFC_guiana))) %>%
    add_column(., Sig = ifelse(.$Gene %in% PmainLFC_guiana_sig$Gene, "Significant", "Not Significant")),
  
  add_column(PmainLFC_iquitos, Group = rep("Iquitos", nrow(PmainLFC_iquitos))) %>%
     add_column(., Sig = ifelse(.$Gene %in% PmainLFC_iquitos_sig$Gene, "Significant", "Not Significant")),
  
  add_column(PmainLFC_maranon, Group = rep("Marañón", nrow(PmainLFC_maranon))) %>%
     add_column(., Sig = ifelse(.$Gene %in% PmainLFC_maranon_sig$Gene, "Significant", "Not Significant")),

  add_column(PmainLFC_nanay, Group = rep("Nanay", nrow(PmainLFC_nanay))) %>%
     add_column(., Sig = ifelse(.$Gene %in% PmainLFC_nanay_sig$Gene, "Significant", "Not Significant"))
) %>%
  mutate(., Gene = gsub("\\.1", "", Gene))

# 5-prime correlations
lfc_pbs_df <- data.frame()
for (i in unique(PmainLFC_combGroups$Group)){
  tmp1 <- pbs_df %>%
    subset(., Group == i & Type == "up_pbs") 
  
  tmp2 <- PmainLFC_combGroups  %>%
    subset(., Group == i) %>%
    dplyr::select(., -Group)
  
  tmp3 <- inner_join(tmp1, tmp2, by = "Gene")
  
  lfc_pbs_df <- rbind(lfc_pbs_df, tmp3)
  
}

pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/5prime_5KbPBS_sigGenes_lfcCorr.pdf", height = 12, width = 6)
my.formula = y ~ x
ggplot(lfc_pbs_df, aes(x = PBS, y = log2FoldChange))  +
  geom_point(data = subset(lfc_pbs_df, Sig == "Significant"), aes(color = Group), alpha = 0.5) +
  geom_point(data = subset(lfc_pbs_df, Sig == "Not Significant"), alpha = 0.1, color = "black") +
  #scale_alpha_discrete(range = c(0.1,1)) + 
  theme_minimal_grid() + 
  # scale_color_manual(values = c( c("grey", "steelblue"), c("grey", "firebrick"),
  #                              c("grey", "darkgreen"), c("grey", "darkorange"))) +
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
  geom_smooth(linetype = 3, color = "red") +
  stat_poly_eq(formula = my.formula, 
               aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               parse = TRUE) +
  theme(
    legend.position = "none"
  ) + 
  facet_wrap(~ Group, ncol = 1, nrow = 4, )
dev.off()

# 3-prime correlations
lfc_pbs_df <- data.frame()
for (i in unique(PmainLFC_combGroups$Group)){
  tmp1 <- pbs_df %>%
    subset(., Group == i & Type == "down_pbs") 
  
  tmp2 <- PmainLFC_combGroups  %>%
    subset(., Group == i) %>%
    dplyr::select(., -Group)
  
  tmp3 <- inner_join(tmp1, tmp2, by = "Gene")
  
  lfc_pbs_df <- rbind(lfc_pbs_df, tmp3)
  
}

pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/3prime_5KbPBS_sigGenes_lfcCorr.pdf", height = 12, width = 6)
my.formula = y ~ x
ggplot(lfc_pbs_df, aes(x = PBS, y = log2FoldChange))  +
  geom_point(data = subset(lfc_pbs_df, Sig == "Significant"), aes(color = Group), alpha = 0.5) +
  geom_point(data = subset(lfc_pbs_df, Sig == "Not Significant"), alpha = 0.1, color = "black") +
  #scale_alpha_discrete(range = c(0.1,1)) + 
  theme_minimal_grid() + 
  # scale_color_manual(values = c( c("grey", "steelblue"), c("grey", "firebrick"),
  #                              c("grey", "darkgreen"), c("grey", "darkorange"))) +
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
  geom_smooth(linetype = 3, color = "red") +
  stat_poly_eq(formula = my.formula, 
               aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")), 
               parse = TRUE) +
  theme(
    legend.position = "none"
  ) + 
  facet_wrap(~ Group, ncol = 1, nrow = 4, )
dev.off()

```


#### PLOT -- venn overlap of top 1% 5-prime PBS genes and top 1000 LFC genes from each genetic group x contrast
```{r}
pbs_5Kb_top1 <- subset(pbs_df_top1, Type == "up_pbs") %>%
  mutate(., Gene = gsub("$", "\\.1", Gene))

pbs_df_top1 <- pbs_df_top1 %>% 
  mutate(., Gene = gsub("$", "\\.1", Gene))

pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/PBS_DEG_overlapVenn.pdf")
par(mfrow=c(4,1))

# Guiana
ggvenn(
  list("Guiana \nTreatment" = as.character(TmainLFC_guiana_lfc1000$Gene),
       "Guiana \nPhenotype" = as.character(PmainLFC_guiana_lfc1000$Gene),
       "Guiana \nTop 1% PBS" = as.character(subset(pbs_df_top1, Group == "Guiana")$Gene)), 
  fill_color = c("lightgrey", "lightgrey", "steelblue"),
  stroke_size = 0.5, set_name_size = 7, 
  show_percentage = FALSE, fill_alpha = 0.8, 
  text_size = 7
  )

# Iquitos
ggvenn(
  list("Iquitos \nTreatment" = as.character(TmainLFC_iquitos_lfc1000$Gene),
       "Iquitos \nPhenotype" = as.character(PmainLFC_iquitos_lfc1000$Gene),
       "Iquitos \nTop 1% PBS" = as.character(subset(pbs_df_top1, Group == "Iquitos")$Gene)), 
  fill_color = c("lightgrey", "lightgrey", "firebrick"),
  stroke_size = 0.5, set_name_size = 7, 
  show_percentage = FALSE, fill_alpha = 0.8,
  text_size = 7
  )

# Maranon
ggvenn(
  list("Marañón \nTreatment" = as.character(TmainLFC_maranon_lfc1000$Gene),
       "Marañón \nPhenotype" = as.character(PmainLFC_maranon_lfc1000$Gene),
       "Marañón \nTop 1% PBS" = as.character(subset(pbs_df_top1, Group == "Marañón")$Gene)), 
  fill_color = c("lightgrey", "lightgrey", "darkgreen"),
  stroke_size = 0.5, set_name_size = 7, 
  show_percentage = FALSE, fill_alpha = 0.8,
  text_size = 7
  )

# Nanay
ggvenn(
  list("Nanay \nTreatment" = as.character(TmainLFC_nanay_lfc1000$Gene),
       "Nanay \nPhenotype" = as.character(PmainLFC_nanay_lfc1000$Gene),
       "Nanay \nTop 1% PBS" = as.character(subset(pbs_df_top1, Group == "Nanay")$Gene)), 
  fill_color = c("lightgrey", "lightgrey", "darkorange"),
  stroke_size = 0.5, set_name_size = 7, 
  show_percentage = FALSE, fill_alpha = 0.8,
  text_size = 7
  )
dev.off()

library(VennDiagram)
# Make Venn diagram from list of groups
set1 <- sample(1:10000, 2000)
set2 <- sample(1:10000, 2000)
set3 <- sample(1:10000, 2000)
set4 <- sample(1:10000, 2000)
colors <- c("darkgrey","darkgrey","darkgrey","darkgrey")
venn.diagram(x = list(set1, set2, set3, set4),
            category.names = c("Guiana", "Iquitos", "Marañón", "Nanay"),
            filename = '~/Downloads/test_venn.png',
            output=TRUE,
            imagetype="png", 
            scaled = FALSE,
            col = "black",
            fill = colors,
            #cat.col = colors,
            cat.cex = 2,
            main.cex = 0,
            sub.cex = 0,
            margin = 0.15,
            cex = 0, 
            alpha = 0.9
)

```

#### PLOT -- venn overlap intersection -- top 1% 5-prime PBS genes and top 1000 LFC genes from each genetic group x contrast
```{r}

lfc_pbs_topTerms <- rbind(
  # Guiana
  rbind(
    attr(venn(list("Treatment" = as.character(TmainLFC_guiana_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_guiana_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Guiana")$Gene))), 
         "intersections")$'Treatment:Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Guiana", nrow(.)),
                 Contrast = rep("Both", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_guiana_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_guiana_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Guiana")$Gene))), 
         "intersections")$'Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Guiana", nrow(.)),
                 Contrast = rep("Phenotype", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_guiana_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_guiana_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Guiana")$Gene))), 
         "intersections")$'Treatment:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Guiana", nrow(.)),
                 Contrast = rep("Treatment", nrow(.)))
  ),
    
  # Iquitos
  rbind(
    # Guiana
    attr(venn(list("Treatment" = as.character(TmainLFC_iquitos_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_iquitos_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Iquitos")$Gene))), 
         "intersections")$'Treatment:Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Iquitos", nrow(.)),
                 Contrast = rep("Both", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_iquitos_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_iquitos_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Iquitos")$Gene))), 
         "intersections")$'Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Iquitos", nrow(.)),
                 Contrast = rep("Phenotype", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_iquitos_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_iquitos_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Guiana")$Gene))), 
         "intersections")$'Treatment:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Iquitos", nrow(.)),
                 Contrast = rep("Treatment", nrow(.)))
  ),
  
  rbind(
    # Marañón
    attr(venn(list("Treatment" = as.character(TmainLFC_maranon_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_maranon_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Marañón")$Gene))), 
         "intersections")$'Treatment:Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Marañón", nrow(.)),
                 Contrast = rep("Both", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_maranon_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_maranon_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Marañón")$Gene))), 
         "intersections")$'Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Marañón", nrow(.)),
                 Contrast = rep("Phenotype", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_maranon_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_maranon_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Marañón")$Gene))), 
         "intersections")$'Treatment:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Marañón", nrow(.)),
                 Contrast = rep("Treatment", nrow(.)))
  ),
  
  rbind(
    # Nanay
    attr(venn(list("Treatment" = as.character(TmainLFC_nanay_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_nanay_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Nanay")$Gene))), 
         "intersections")$'Treatment:Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Nanay", nrow(.)),
                 Contrast = rep("Both", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_nanay_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_nanay_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Nanay")$Gene))), 
         "intersections")$'Phenotype:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Nanay", nrow(.)),
                 Contrast = rep("Phenotype", nrow(.))),
    
    attr(venn(list("Treatment" = as.character(TmainLFC_nanay_lfc1000$Gene),
                   "Phenotype" = as.character(PmainLFC_nanay_lfc1000$Gene),
                   "PBS" = as.character(subset(pbs_df_top1, Group == "Nanay")$Gene))), 
         "intersections")$'Treatment:PBS' %>%
      data.frame(.) %>% 
      setNames(., nm = "Gene") %>%
      inner_join(., annot, by = "Gene") %>%
      add_column(Population = rep("Nanay", nrow(.)),
                 Contrast = rep("Treatment", nrow(.)))
  )
)

# Write out tables for supplemental materials
# Top 1% PBS terms that intersect with differential expression results
write.table(lfc_pbs_topTerms, "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/TablsS7_PBS_overlapDiffExprGenes.tsv", row.names = FALSE, quote = FALSE, sep = '\t')
# Top 1% PBS terms--ALL
write.table(pbs_df_top1, "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/TablsS8_PBS_top1.tsv", row.names = FALSE, quote = FALSE, sep = '\t')


x <- left_join(lfc_pbs_topTerms, shared_genes_both_consGOTerms, by = "Gene")

```

#### PLOT -- Manhattan plots from PBS analyses
```{r}
# Need gene positions to organizse the SNPs
gene_pos <- read.table("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/SCA-6_genePos.txt", header = FALSE) %>% 
  setNames(nm = c("Chr", "Gene", "Start", "Stop")) %>%
  # Extend genes by 5kb on either side (what Tuomas did in his PBS analysis)
  # Note: these are not exact positions, but work well enough for plotting
  mutate(Start = Start - 5000) %>% 
  mutate(Start = ifelse(Start < 0, 0, Start)) %>% 
  mutate(Stop = Stop + 5000) %>% 
  mutate(Midpoint = round((Start + Stop) / 2)) %>% 
  mutate(Chr = gsub("Chr", "", Chr))

# Chromosome positions
chr_pos <- 
  gene_pos %>% 
  mutate(Chr = gsub("Chr", "", Chr)) %>% 
  mutate(Chr = as.numeric(as.character(Chr)))
chr_pos <- chr_pos[order(chr_pos$Chr),]
chr_pos <- 
  chr_pos %>%
  group_by(Chr) %>%
  summarise(chr_len=max(Midpoint)) %>%
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  dplyr::select(-chr_len) %>%
  mutate(Chr = as.character(Chr)) %>%
  left_join(gene_pos, by = "Chr") %>%
  mutate( BPcum=Midpoint+tot)
  

# For Phenotype
lfc_pbs_df <- data.frame()
for (i in unique(PmainLFC_combGroups$Group)){
  tmp1 <- pbs_df %>%
    subset(., Group == i) 
  
  tmp2 <- PmainLFC_combGroups  %>%
    subset(., Group == i) %>%
    dplyr::select(., -Group)
  
  tmp3 <- inner_join(tmp1, tmp2, by = "Gene")
  
  lfc_pbs_df <- rbind(lfc_pbs_df, tmp3)
  
}


lfc_pbs_df <- 
  lfc_pbs_df %>% 
  inner_join(chr_pos, by = "Gene") %>%
  subset(Chr != "0") %>%
  mutate(Chr = as.numeric(as.character(Chr))) %>% 
  #mutate(top1 = ifelse(Gene %in% pbs_df_top1$Gene, "Top 1% PBS Scores", "Not Top 1% PBS Scores"))
lfc_pbs_df <- 
  lfc_pbs_df[order(lfc_pbs_df$Chr),] 


axisdf = chr_pos %>%
  group_by(Chr) %>% 
  summarise(center=( max(BPcum) + min(BPcum) ) / 2 ) %>% 
  subset(Chr != 0) %>%
  mutate(Chr = as.numeric(as.character(Chr))) 
axisdf <- axisdf[order(axisdf$Chr),]
  
pbs_hlines <- 
  pbs_df_top1 %>% 
  group_by(Group, Type) %>%
  summarise(hline = min(PBS)) %>% 
  subset(Group == "Guiana")
  

pdf("~/Downloads/test.pdf", height = 4, width = 4)
lfc_pbs_df %>% 
  subset(Group == "Guiana") %>%
  mutate(color = ifelse(Sig == "Significant" & top1 == "Top 1% PBS Scores", "color", "no_color")) %>%
  ggplot(aes(x = Chr, y = PBS, color = color)) + 
  geom_jitter(alpha = 0.6, width = 0.3) +
  scale_color_manual(values = c("red", "grey")) +
  facet_grid(Group ~ Type) + 
  scale_x_discrete( limits = axisdf$Chr) + 
  geom_hline(data = pbs_hlines, aes(yintercept = hline))
dev.off()

  
```


# Structural variants overlapping DEGs
```{r}
# Criollo x SCA-6 collinear blocks as determined using MCScanX
collin <- read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Structural_Variants/SCA6_vs_Criollo_MCScanX/SCA-6_vs_Criollo.tab.collinearity.csv", header = TRUE) %>% 
  rename(source = "Criollo_Gene", target = "SCA6_Gene")

# Selection outliers from Hamala & Wafula et al., 2021--Supplemental table S4
sel_outliers <- read.csv("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Structural_Variants/SV_TableS4_svSelOutliers.csv", header = TRUE) %>% 
  rename(Gene = "Criollo_Gene")

x <- inner_join(collin, sel_outliers, by = "Criollo_Gene")

collin_sel_outliers_treat <- 
  inner_join(collin, sel_outliers, by = "Criollo_Gene") %>%
  dplyr::select(SCA6_Gene) %>% 
  unique() %>% 
  rename(SCA6_Gene = "Gene") %>% 
  mutate(Gene = gsub("$", "\\.1", Gene)) %>%
  inner_join(TmainLFC_all_pop_lfc1000, by = "Gene")

collin_sel_outliers_treat <- 
  TmainLFC_all_pop_lfc1000 %>% 
  add_column(SV = ifelse(.$Gene %in% collin_sel_outliers_treat$Gene, "SV_Outlier", "No_SV"))

ggplot(collin_sel_outliers_treat, aes(x = SV, y = log2FoldChange, color = SV)) +
  geom_boxplot() + 
  facet_wrap(~Population)

attr(venn(list(
  Guiana = as.character(subset(collin_sel_outliers_treat, Population == "Guiana")$Gene),
  Iquitos = as.character(subset(collin_sel_outliers_treat, Population == "Iquitos")$Gene),
  Marañón = as.character(subset(collin_sel_outliers_treat, Population == "Marañón")$Gene),
  Nanay = as.character(subset(collin_sel_outliers_treat, Population == "Nanay")$Gene)
)), "intersections")

collin_sel_outliers_pheno <- 
  inner_join(collin, sel_outliers, by = "Criollo_Gene") %>%
  dplyr::select(SCA6_Gene) %>% 
  unique() %>% 
  rename(SCA6_Gene = "Gene") %>% 
  mutate(Gene = gsub("$", "\\.1", Gene))

collin_sel_outliers_pheno <- 
  PmainLFC_all_pop_lfc1000 %>%
  add_column(SV = ifelse(.$Gene %in% collin_sel_outliers_pheno$Gene, "SV_Outlier", "No_SV"))

ggplot(collin_sel_outliers_pheno, aes(x = SV, y = log2FoldChange, color = SV)) +
  geom_boxplot() + 
  facet_wrap(~Population)

attr(venn(list(
  Guiana = as.character(subset(collin_sel_outliers_pheno, Population == "Guiana")$Gene),
  Iquitos = as.character(subset(collin_sel_outliers_pheno, Population == "Iquitos")$Gene),
  Marañón = as.character(subset(collin_sel_outliers_pheno, Population == "Marañón")$Gene),
  Nanay = as.character(subset(collin_sel_outliers_pheno, Population == "Nanay")$Gene)
)), "intersections")

nrow(subset(PmainLFC_all_pop, padj < 0.05 & Population == "Nanay"))

```







##################
### SUPPLEMENT ###
##################
## Calculation and plots for Supplemental figures
#### Suppemental figure, boxplot of enrichedGOExpresisonDifferences
```{r}
#load("~/Desktop/NSF_modelSelection/topGO_plots_outline_lfcEstimates.rdata")
tmp_df_grp <- 
  norm_expr[rownames(norm_expr) %in% all_go_lfc$Gene,] %>%
  t(.) %>%
  data.frame(.) %>%
  rownames_to_column(., var = "Library_ID") %>%
  inner_join(rownames_to_column(colData, var = "Library_ID"), .,  by = "Library_ID") %>%
  dplyr::select(., c(Genotype, Gen_Group, Treatment, Phenotype),  starts_with("SCA.6")) %>%
  pivot_longer(., cols = starts_with("SCA.6"), names_to = "Gene", values_to = "Expression") %>%
  group_by(Gen_Group, Gene, Treatment) %>%
  summarize(meanExpr = mean(Expression)) %>%
  mutate(., Gene = gsub("SCA\\.6", "SCA-6", Gene))

tmp_df_full <- 
  norm_expr[rownames(norm_expr) %in% all_go_lfc$Gene,] %>%
  t(.) %>%
  data.frame(.) %>%
  rownames_to_column(., var = "Library_ID") %>%
  inner_join(rownames_to_column(colData, var = "Library_ID"), .,  by = "Library_ID") %>%
  dplyr::select(., c(Genotype, Gen_Group, Treatment, Phenotype),  starts_with("SCA.6")) %>%
  mutate(Gen_Group = rep("Full", nrow(.))) %>%
  pivot_longer(., cols = starts_with("SCA.6"), names_to = "Gene", values_to = "Expression") %>%
  group_by(Gen_Group, Gene, Treatment) %>%
  summarize(meanExpr = mean(Expression)) %>%
  mutate(., Gene = gsub("SCA\\.6", "SCA-6", Gene))

tmp_df_grp_full <- rbind(tmp_df_full, tmp_df_grp)

tmp_df_grp_full <- all_go_lfc %>%
  dplyr::select(., c(Gene, GO.ID, Term)) %>%
  inner_join(., tmp_df_grp_full, by = "Gene") %>%
  mutate(., Gen_Group = ifelse(.$Gen_Group == "Maranon", "Marañón", .$Gen_Group))

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_boxplot_enrichedGOExpression.pdf",
    height = 35,
    width = 20)
ggplot(tmp_df_grp_full, aes(x = Gen_Group, y = meanExpr, fill = Gen_Group)) + 
  geom_violin(aes(group = interaction(Treatment, Gen_Group)), alpha = 0.7) +
  geom_boxplot(aes(group = interaction(Treatment, Gen_Group)), width = 0.5, position = position_dodge(width = 0.9)) +
  scale_fill_manual(values = c("black", "steelblue", "firebrick", "darkgreen", "darkorange")) + 
  labs(y = "Mean Normalized Expression") +
  theme_minimal_grid() +
  facet_wrap(~ Term, ncol = 2, nrow = 5) +
  theme(strip.placement =  "inside",
        plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(size = 30, face = "bold"),
        #axis.ticks.x = element_text(size = 30, face = "bold"),
        strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
        strip.text.x = element_text(size = 30, face = "bold"),
        strip.text.y = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 30, face = "bold"),
        axis.title.y = element_text(size = 30, face = "bold"),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 30),
        legend.title = element_text(size = 30),
        legend.position = "bottom")
dev.off()

```

### Supplemental figure, cluster LFC for each gene x genetic group to look for population-specific expression clusters 
```{r}
#######################################
### K-means clustering -- TREATMENT ###
#######################################
top_genes_eachPop <- 
  unique(c(TmainLFC_guiana_lfc1000$Gene,
           TmainLFC_iquitos_lfc1000$Gene,
           TmainLFC_maranon_lfc1000$Gene,
           TmainLFC_nanay_lfc1000$Gene))
top_genes_eachPop <- trt_log2FC_allPop[trt_log2FC_allPop$Gene %in% top_genes_eachPop,] %>%
  remove_rownames(.) %>%
  column_to_rownames(., var = "Gene") %>%
  dplyr::select(., -Full)

withinSSE <- as.numeric()

# SSE criterion
# SSE is the SS distance between a point and its cluster centroid, can calculate across range of k
# At a certain value k, SSE will not sig decrease with each new addition of a cluster
# Looking at the "elbow" of the plot, we see this value is 6
for (i in 2:30) withinSSE[i] <- sum(kmeans(top_genes_eachPop, centers=i)$withinss)

# Calinski criterion
# Essentially uses the intra vs inter SSE for each cluster and uses it to create an index
# Maximizing this index indicates your cluster have a low intra-cluster SSE and a high inter-cluster SSE
# Looking at the highest index value, highlighted in red, we see this value is 6
fit <- cascadeKM(top_genes_eachPop, 1, 30, iter = 5)

# Silhouette width
sil <- rep(0, 20)
#repeat k-means for 1:20 and extract silhouette:
for(i in 2:20){
  k1to20 <- kmeans(top_genes_eachPop, centers = i, nstart = 25, iter.max = 50)
  ss <- silhouette(k1to20$cluster, dist(top_genes_eachPop))
  sil[i] <- mean(ss[, 3])
}

#### PLOT -- Supplemental figure, population-specific expression clusters -- cluster size determination
pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/kmeans_clusterDeter_treatment.pdf")
# SSE
plot(1:30, withinSSE, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares")

# CH Criterion
plot(fit, sortg = TRUE, grpmts.plot = TRUE)

# Plot the  average silhouette width
plot(1:20, sil, type = "b", pch = 19, xlab = "Number of clusters k", ylab="Average silhouette width")
abline(v = which.max(sil), lty = 2)

dev.off()

#### PLOT -- Supplemental figure, population-specific expression clusters
# k-means clustering with n = 5 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 5)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p1 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )

# k-means clustering with n = 10 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 10)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p2 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )
  
 
# k-means clustering with n = 15 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 15)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p3 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )
   

# k-means clustering with n = 20 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 20)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p4 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )
    

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/popSpecific_effSizeClusters_treatment.pdf", width = 20, height = 20)
plot(p1)
plot(p2)
plot(p3)
plot(p4)
dev.off()

#######################################
### K-means clustering -- PHENOTYPE ###
#######################################
top_genes_eachPop <- 
  unique(c(PmainLFC_guiana_lfc1000$Gene,
           PmainLFC_iquitos_lfc1000$Gene,
           PmainLFC_maranon_lfc1000$Gene,
           PmainLFC_nanay_lfc1000$Gene))
top_genes_eachPop <- phe_log2FC_allPop[phe_log2FC_allPop$Gene %in% top_genes_eachPop,] %>%
  remove_rownames(.) %>%
  column_to_rownames(., var = "Gene") %>%
  dplyr::select(., -Full)

withinSSE <- as.numeric()

# SSE criterion
# SSE is the SS distance between a point and its cluster centroid, can calculate across range of k
# At a certain value k, SSE will not sig decrease with each new addition of a cluster
# Looking at the "elbow" of the plot, we see this value is 6
for (i in 2:30) withinSSE[i] <- sum(kmeans(top_genes_eachPop, centers=i)$withinss)

# Calinski criterion
# Essentially uses the intra vs inter SSE for each cluster and uses it to create an index
# Maximizing this index indicates your cluster have a low intra-cluster SSE and a high inter-cluster SSE
# Looking at the highest index value, highlighted in red, we see this value is 6
fit <- cascadeKM(top_genes_eachPop, 1, 30, iter = 5)

# Silhouette width
sil <- rep(0, 20)
#repeat k-means for 1:20 and extract silhouette:
for(i in 2:20){
  k1to20 <- kmeans(top_genes_eachPop, centers = i, nstart = 25, iter.max = 50)
  ss <- silhouette(k1to20$cluster, dist(top_genes_eachPop))
  sil[i] <- mean(ss[, 3])
}

#### PLOT -- Supplemental figure, population-specific expression clusters -- cluster size determination
pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/kmeans_clusterDeter_phenotype.pdf")
# SSE
plot(1:30, withinSSE, type="b", xlab="Number of Clusters",
     ylab="Within groups sum of squares")

# CH Criterion
plot(fit, sortg = TRUE, grpmts.plot = TRUE)

# Plot the  average silhouette width
plot(1:20, sil, type = "b", pch = 19, xlab = "Number of clusters k", ylab="Average silhouette width")
abline(v = which.max(sil), lty = 2)

dev.off()

#### PLOT -- Supplemental figure, population-specific expression clusters
# k-means clustering with n = 5 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 5)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p1 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )

# k-means clustering with n = 10 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 10)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p2 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )
  
 
# k-means clustering with n = 15 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 15)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p3 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )
   

# k-means clustering with n = 20 clusters
kmeans_effSize <- kmeans(top_genes_eachPop, centers = 20)
clusterEffSize <- data.frame(kmeans_effSize$cluster) %>%
    rownames_to_column(., var = "Gene") %>%
    setNames(., nm = c("Gene", "Cluster")) %>% 
    inner_join(., rownames_to_column(top_genes_eachPop, var = "Gene"), by = "Gene") %>%
    melt(., id.vars = c("Gene", "Cluster"), variable.name = "Population", value.name = "log2FoldChange")
p4 <-
    ggplot(clusterEffSize, aes(x = Population, y = log2FoldChange, color = Population)) + 
      geom_boxplot(width = 0.35) + 
      scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
      #scale_y_continuous(breaks = seq(-20,20, 2)) +
      facet_wrap(~Cluster) +
      theme(
        axis.text.x = element_text(angle = 45, vjust = 0.5)
    )

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/popSpecific_effSizeClusters_phenotype.pdf", width = 20, height = 20)
plot(p1)
plot(p2)
plot(p3)
plot(p4)
dev.off()

```


#### Supplemental figure, LFC for various # gene subsets (100-1000) each population
```{r}
#################
### TREATMENT ###
#################
med_eff <- seq(100,1000, 100)
genetic_group <- c("Guiana", "Iquitos", "Marañón", "Nanay")
tmp_med <- data.frame()
stderr <- function(x) sd(x)/sqrt(length(x))

for (k in genetic_group){
  tmp_df <- trt_log2FC_allPop[,c("Gene", k)] %>% 
    setNames(., c("Gene", "log2FC")) %>%
    mutate(absLog2FC = abs(log2FC))
  
  for (i in med_eff){
  
    y <- tmp_df[order(tmp_df$abs, decreasing = TRUE),] %>%
      .[1:i,] %>%
      dplyr::select(., "Gene") %>%
      inner_join(., tmp_df, by = "Gene")
    z <- anti_join(tmp_df, dplyr::select(y, "Gene"), by = "Gene")
    
    y <- c(median(y$absLog2FC), stderr(y$absLog2FC))
    z <- c(median(z$absLog2FC), stderr(z$absLog2FC))
    
    tmp_med <- rbind(tmp_med, c(y, z))
  
  }
  
}

tmp_med$Pop <- rep(genetic_group, each = 10)
tmp_med$n <- rep(med_eff, 4)
colnames(tmp_med) <- c("median_top","stdev_top", "median_remaining", "stdev_remaining", "Population", "N")

tmp_med <- cbind(
  melt(tmp_med, id.var = c("Population","N"), measure.vars = c("median_top", "median_remaining"), variable.name = "Median", value.name = "Median_absLog2FoldChange"),
  
  melt(tmp_med, id.var = c("Population","N"), measure.vars = c("stdev_top", "stdev_remaining"), variable.name = "StDev", value.name = "StDev_absLog2FoldChange") %>%
    dplyr::select(., c(StDev, StDev_absLog2FoldChange))
          ) 
pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/topGene_LFC_perGroup_treatment.pdf",
    height = 30, 
    width = 50)           
ggplot(tmp_med, aes( x = Median, y = Median_absLog2FoldChange, color = Median)) +
  geom_point(size = 8) +
  geom_errorbar(stat = "identity", 
                aes(ymin = Median_absLog2FoldChange - StDev_absLog2FoldChange,
                ymax = Median_absLog2FoldChange + StDev_absLog2FoldChange), size = 3) +
  scale_x_discrete(labels = c("median_top" = "Top", "median_remaining" = "Remaining")) +
  scale_color_manual(values = c("#edae49", "#2e4057")) +
  facet_wrap(~Population + N, scales = "free_x", nrow = 4, ncol = 10) + 
  labs(x = NULL,
       y = "Median | log2FoldChange |",
       title = "Median LFC for different sized cross sections -- Treatment") +
  theme(axis.title.y =  element_text(size = 30, face = "bold"),
        axis.text.x =  element_text(size = 30, face = "bold"),
        strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
        strip.text.x = element_text(size = 30, face = "bold"),
        strip.text.y = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 30, face = "bold"),
        title = element_text(size = 30, face = "bold"))
dev.off()


#################
### PHENOTYPE ###
#################
med_eff <- seq(100,1000, 100)
genetic_group <- c("Guiana", "Iquitos", "Marañón", "Nanay")
tmp_med <- data.frame()
stderr <- function(x) sd(x)/sqrt(length(x))

for (k in genetic_group){
  tmp_df <- phe_log2FC_allPop[,c("Gene", k)] %>% 
    setNames(., c("Gene", "log2FC")) %>%
    mutate(absLog2FC = abs(log2FC))
  
  for (i in med_eff){
  
    y <- tmp_df[order(tmp_df$abs, decreasing = TRUE),] %>%
      .[1:i,] %>%
      dplyr::select(., "Gene") %>%
      inner_join(., tmp_df, by = "Gene")
    z <- anti_join(tmp_df, dplyr::select(y, "Gene"), by = "Gene")
    
    y <- c(median(y$absLog2FC), stderr(y$absLog2FC))
    z <- c(median(z$absLog2FC), stderr(z$absLog2FC))
    
    tmp_med <- rbind(tmp_med, c(y, z))
  
  }
  
}

tmp_med$Pop <- rep(genetic_group, each = 10)
tmp_med$n <- rep(med_eff, 4)
colnames(tmp_med) <- c("median_top","stdev_top", "median_remaining", "stdev_remaining", "Population", "N")

tmp_med <- cbind(
  melt(tmp_med, id.var = c("Population","N"), measure.vars = c("median_top", "median_remaining"), variable.name = "Median", value.name = "Median_absLog2FoldChange"),
  
  melt(tmp_med, id.var = c("Population","N"), measure.vars = c("stdev_top", "stdev_remaining"), variable.name = "StDev", value.name = "StDev_absLog2FoldChange") %>%
    dplyr::select(., c(StDev, StDev_absLog2FoldChange))
          ) 
pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/topGene_LFC_perGroup_phenotype.pdf",
    height = 30, 
    width = 50)           
ggplot(tmp_med, aes( x = Median, y = Median_absLog2FoldChange, color = Median)) +
  geom_point(size = 8) +
  geom_errorbar(stat = "identity", 
                aes(ymin = Median_absLog2FoldChange - StDev_absLog2FoldChange,
                ymax = Median_absLog2FoldChange + StDev_absLog2FoldChange), size = 3) +
  scale_x_discrete(labels = c("median_top" = "Top", "median_remaining" = "Remaining")) +
  scale_color_manual(values = c("#edae49", "#2e4057")) +
  facet_wrap(~Population + N, scales = "free_x", nrow = 4, ncol = 10) + 
  labs(x = NULL,
       y = "Median | log2FoldChange |", 
       title = "Median LFC for different sized cross sections -- Phenotype") +
  theme(axis.title.y =  element_text(size = 30, face = "bold"),
        axis.text.x =  element_text(size = 30, face = "bold"),
        strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
        strip.text.x = element_text(size = 30, face = "bold"),
        strip.text.y = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 30, face = "bold"),
        title = element_text(size = 30, face = "bold"))
dev.off()



```

### Supplemental Figure--Enrichments for PBS genes
```{r}
#####################
### PBS Genes--GO ###
#####################

ont <- c("BP", "MF")
pbs_pos <- c("up_pbs", "coding_pbs", "down_pbs")
genetic_group <- c("Guiana", "Iquitos", "Maranon", "Nanay")

# Load geneID2GO file using readMappings function 
geneIDGO <- readMappings(file = "~/Desktop/NSF_modelSelection/SCA6_GO.txt", sep = '\t', IDsep = ';')

# How do you want to define significance--here it is genes with padj < 0.05, could choose LFC cutoff
topDiffGenesSig <- function(allScore) {return(abs(allScore) > 0)}
 
# Ugly nested for loop to iterate over each PBS position, genetic group, and ontology 
for (i in pbs_pos){
  for(k in genetic_group){
    for(j in ont){
      
      # Subset the dataframe
      tmp_pbs <- subset(pbs_df_top1, Type == i & Group == k)
    
      # All expressed genes
      allExprGenes <- pbs_df %>%
        subset(., Type == i & Group == k) %>%
        mutate(Sig = ifelse(.$Gene %in% tmp_pbs$Gene, 1, 0)) %>%
        dplyr::select(., c(Gene, Sig)) %>%
        mutate(., Gene = gsub("$", "\\.1", .$Gene)) %>%
        deframe(.)
      
      # Create the topGO object--BP
      GOdata <- new("topGOdata",
                    description = "Simple session", ontology = j,
                    allGenes = allExprGenes, geneSel = topDiffGenesSig,
                    nodeSize = 10,
                    annot = annFUN.gene2GO, 
                    gene2GO = geneIDGO)
        
      # Run the enrichment tests, correct p-values 
      resultClassic <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
      topGO <- GenTable(GOdata, pval = resultClassic, topNodes = length(resultClassic@score))
      topGO$padj <- p.adjust(topGO$pval, method = "fdr")
      topGO$ontology <- rep(j, nrow(topGO))
      
      assign(paste0("topGO_", j), topGO)
    }
    
    tmp_comb <- rbind(topGO_BP, topGO_MF)
    assign(paste0("topGO_", k , "_", i), tmp_comb)
    
  }
}  


```

### Rank using GOxploreR
```{r}
go_dfs <- list(topGO_Guiana_up_pbs, topGO_Guiana_coding_pbs, topGO_Guiana_down_pbs,
               topGO_Iquitos_up_pbs, topGO_Iquitos_coding_pbs, topGO_Iquitos_down_pbs,
               topGO_Maranon_up_pbs, topGO_Maranon_coding_pbs, topGO_Maranon_down_pbs,
               topGO_Nanay_up_pbs, topGO_Nanay_coding_pbs, topGO_Nanay_down_pbs)

names(go_dfs) <- c("topGO_Guiana_up_pbs", "topGO_Guiana_coding_pbs", "topGO_Guiana_down_pbs",
                   "topGO_Iquitos_up_pbs", "topGO_Iquitos_coding_pbs", "topGO_Iquitos_down_pbs",
                   "topGO_Maranon_up_pbs", "topGO_Maranon_coding_pbs", "topGO_Maranon_down_pbs",
                   "topGO_Nanay_up_pbs", "topGO_Nanay_coding_pbs", "topGO_Nanay_down_pbs")

for (k in 1:length(go_dfs)){
  terms <- go_dfs[[k]]
  terms <- subset(terms, ontology == "BP" & padj < 0.05)
  slim_id <- prioritizedGOTerms(as.character(terms$GO.ID), organism = "Cress", sp=TRUE, domain = "BP")$HF
  slimGOTerms <- terms[terms$GO.ID %in% slim_id,]
  
  nm <- gsub("topGO_", "", names(go_dfs)[k])
  slimGOTerms$Group_Pos <- rep(nm, nrow(slimGOTerms))
  set_name <- paste0(names(go_dfs)[k], "_slimGOTerms")
  print(set_name)
  assign(set_name, slimGOTerms)
}


```

#### Supplemental Figure-- BinGO style plot of enriched GO terms for each group's PBS scores
```{r}
x <- rbind(topGO_Guiana_up_pbs_slimGOTerms, 
           topGO_Guiana_coding_pbs_slimGOTerms, 
           topGO_Guiana_down_pbs_slimGOTerms,
           topGO_Iquitos_up_pbs_slimGOTerms, 
           topGO_Iquitos_coding_pbs_slimGOT8erms, 
           topGO_Iquitos_down_pbs_slimGOTerms,
           topGO_Maranon_up_pbs_slimGOTerms, 
           topGO_Maranon_coding_pbs_slimGOTerms, 
           topGO_Maranon_down_pbs_slimGOTerms,
           topGO_Nanay_up_pbs_slimGOTerms, 
           topGO_Nanay_coding_pbs_slimGOTerms, 
           topGO_Nanay_down_pbs_slimGOTerms) 
  
y <- 
  got %>%
  setNames(., nm = c("Gene","GO.ID")) %>%
  inner_join(., x, by = "GO.ID") %>%
  add_column(Group_Pos_Gene = paste0(.$Group_Pos, "_", .$Gene)) %>%
  dplyr::select(., c(Group_Pos_Gene, GO.ID))

z <- pbs_df_top1 %>%
  add_column(Group_Pos_Gene = paste0(.$Group, "_", .$Type, "_", .$Gene)) %>%
  inner_join(., y, by = "Group_Pos_Gene") %>%
  dplyr::select(., -Group_Pos_Gene) %>%
  add_column(., Group_Pos = paste0(.$Group, "_", .$Type))

z <- z %>%
  group_by(Group_Pos, GO.ID) %>%
  summarize(median_pbs = median(PBS)) %>%
  inner_join(., unique(dplyr::select(x, c(GO.ID, Term))), by = "GO.ID") %>%
  add_column(., Population = gsub("_.*", "", .$Group_Pos))

pdf("~/Downloads/test.pdf", 
    width = 100,
    height = 100)

z %>%
  ggplot() + 
  geom_point(aes(y=tolower(Term),
             x=1,
             size=median_pbs,
             color = Population), 
             ) + 
  scale_color_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
  scale_x_continuous(limits = c(1,1), breaks = 0) +
  facet_grid(~Group_Pos) + 
  guides(color = "none") + 
  labs(size = "Median PBS") +
  theme(strip.placement = "outside",
        plot.title = element_text(hjust = 0.5),
        axis.title.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.y = element_blank(),
        axis.ticks.x = element_blank(),
        strip.background.y = element_rect(fill = "#EEEEEE", color = "#FFFFFF"),
        strip.text.x = element_text(size = 30, face = "bold"),
        strip.text.y = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 30, face = "bold"),
        legend.text = element_text(size = 30),
        legend.title = element_text(size = 30),
        legend.position = "bottom") +
  scale_size_continuous(range = c(10,25))

dev.off()
  

```


#### Supplemental Figure -- LFC correlation between genes within the GO terms -- shared GO terms
```{r}
# Plot median LFC for each enriched GO term (after GOexploreR and REVIGO filtering)
# LFC correlations
sharedGO_correlations_lfcTrt <- data.frame()
sharedGO_correlations_lfcPhe <- data.frame()

# Treatment
for (t in unique(shared_goTerms_trt_LFC$Term)){
  tmp <- 
    shared_goTerms_trt_LFC %>%
      subset(., Term == t) %>%
      dplyr::select(., Gene) %>%
      unique() %>%
      inner_join(., norm_expr_trt, by = "Gene") %>%
      #pivot_wider(id_cols = Gene, names_from = Population, values_from = log2FoldChange) %>%
      column_to_rownames(., var = "Gene") %>%
      cor(., method = "spearman") 
  
  tmp[upper.tri(tmp)] <- NA
  
  tmp <- as.numeric(na.omit(as.numeric(tmp))) %>%
    .[.<1.000]
  
  sharedGO_correlations_lfcTrt <-
    rbind(sharedGO_correlations_lfcTrt, 
          data.frame(Corr = tmp) %>%
            add_column(Term = rep(t, nrow(.)),
            Contrast = rep("Treatment", nrow(.)))
          )
}

# Phenotype
for (t in unique(shared_goTerms_phe_LFC$Term)){
  tmp <- 
  shared_goTerms_phe_LFC %>%
  subset(., Term == t) %>%
  dplyr::select(., Gene) %>%
  unique() %>%
  inner_join(., TmainLFC_all_pop, by = "Gene") %>%
  pivot_wider(id_cols = Gene, names_from = Population, values_from = log2FoldChange) %>%
  column_to_rownames(., var = "Gene") %>%
  cor(., method = "spearman") 
  
  tmp[upper.tri(tmp)] <- NA
  
  tmp <- as.numeric(na.omit(as.numeric(tmp))) %>%
    .[.<1.000]
  
  sharedGO_correlations_lfcPhe <-
    rbind(sharedGO_correlations_lfcPhe, 
          data.frame(Corr = tmp) %>%
            add_column(Term = rep(t, nrow(.)),
            Contrast = rep("Phenotype", nrow(.))))
}



# Expression correlations
sharedGO_correlations_exprTrt <- data.frame()
sharedGO_correlations_exprPhe <- data.frame()

# Treatment
for (t in unique(shared_goTerms_trt_LFC$Term)){
  tmp <- 
  shared_goTerms_trt_LFC %>%
  subset(., Term == t) %>%
  dplyr::select(., Gene) %>%
  unique() %>%
  inner_join(., norm_expr_trt, by = "Gene") %>%
  column_to_rownames(., var = "Gene") %>%
  cor(., method = "spearman") 
  
  tmp[upper.tri(tmp)] <- NA
  
  tmp <- as.numeric(na.omit(as.numeric(tmp))) %>%
    .[.<1.000]
  
  sharedGO_correlations_exprTrt <-
    rbind(sharedGO_correlations_exprTrt, 
          data.frame(Corr = tmp) %>%
            add_column(Term = rep(t, nrow(.)),
            Contrast = rep("Treatment", nrow(.)))
          )
}

# Phenotype
for (t in unique(shared_goTerms_phe_LFC$Term)){
  tmp <- 
  shared_goTerms_phe_LFC %>%
  subset(., Term == t) %>%
  dplyr::select(., Gene) %>%
  unique() %>%
  inner_join(., norm_expr_phe, by = "Gene") %>%
  column_to_rownames(., var = "Gene") %>%
  cor(., method = "spearman") 
  
  tmp[upper.tri(tmp)] <- NA
  
  tmp <- as.numeric(na.omit(as.numeric(tmp))) %>%
    .[.<1.000]
  
  sharedGO_correlations_exprPhe <-
    rbind(sharedGO_correlations_exprPhe, 
          data.frame(Corr = tmp) %>%
            add_column(Term = rep(t, nrow(.)),
            Contrast = rep("Treatment", nrow(.)))
          )
}

# GGpairs plot
# p2 <- 
#   shared_goTerms_trt_LFC %>%
#   subset(., Term == "lignin biosynthetic process") %>%
#   dplyr::select(., Gene) %>%
#   unique() %>%
#   inner_join(., TmainLFC_all_pop, by = "Gene") %>%
#   pivot_wider(id_cols = Gene, names_from = Population, values_from = log2FoldChange) %>%
#   column_to_rownames(., var = "Gene") %>%
#   ggpairs(., lower = list(continuous = wrap("smooth", alpha = 0.3, size=0.75)), 
#             title = "Lignin Biosynthetic Process") + 
#   theme(axis.text.x = element_text(angle = 45))


pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Fig6_lfcCorr_sharedGOTerms.pdf", width = 20, height = 10)
# plot_grid(ggmatrix_gtable(p1), 
#           ggmatrix_gtable(p2),
#           ggmatrix_gtable(p3),
#           ggmatrix_gtable(p4),
#           ncol = 2, nrow = 2)

# LFC correlation plots
p1 <-
  sharedGO_correlations_lfcTrt %>%
  mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>%
  arrange(desc(Contrast)) %>%
  ggplot(., aes(x = Term, y = Corr)) +
  geom_point(size = 10) + 
  stat_summary(fun = mean, color = "blue", shape = 18, size = 5) +
  labs( y = "Correlation Coefficient (r)", x = NULL) +
  scale_y_continuous(breaks = seq(-1, 1, 0.1)) + 
  theme_minimal_grid() + 
  coord_flip() +
  theme(
    axis.text.x = element_text(size = 20, face = "bold", angle = 45, vjust = 0.6),
    axis.text.y = element_text(size = 30, face = "bold"),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold"), 
  )
p1

p2 <-
  sharedGO_correlations_lfcPhe %>%
  mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>%
  arrange(desc(Contrast)) %>%
  ggplot(., aes(x = Term, y = Corr)) +
  geom_point(size = 10) + 
  stat_summary(fun = mean, color = "blue", shape = 18, size = 5) +
  labs( y = "Correlation Coefficient (r)", x = NULL) +
  scale_y_continuous(breaks = seq(-1, 1, 0.1)) + 
  theme_minimal_grid() + 
  coord_flip() +
  theme(
    axis.text.x = element_text(size = 20, face = "bold", angle = 45, vjust = 0.5),
    axis.text.y = element_text(size = 30, face = "bold"),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold"), 
    panel.spacing = unit(4, "lines")
  )
p2 

# Expression correlation plots
p3 <-
  sharedGO_correlations_exprTrt %>%
  mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>%
  arrange(desc(Contrast)) %>%
  ggplot(., aes(x = Term, y = Corr)) +
  geom_point(size = 10) + 
  stat_summary(fun = mean, color = "blue", shape = 18, size = 5) +
  labs( y = "Correlation Coefficient (r)", x = NULL) +
  scale_y_continuous(breaks = seq(-1, 1, 0.1)) + 
  theme_minimal_grid() + 
  coord_flip() +
  theme(
    axis.text.x = element_text(size = 20, face = "bold", angle = 45, vjust = 0.6),
    axis.text.y = element_text(size = 30, face = "bold"),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold"), 
  )
p3

p4 <-
  sharedGO_correlations_exprPhe %>%
  mutate(Contrast = factor(Contrast, levels = Contrast, labels = Contrast)) %>%
  arrange(desc(Contrast)) %>%
  ggplot(., aes(x = Term, y = Corr)) +
  geom_point(size = 10) + 
  stat_summary(fun = mean, color = "blue", shape = 18, size = 5) +
  labs( y = "Correlation Coefficient (r)", x = NULL) +
  scale_y_continuous(breaks = seq(-1, 1, 0.1)) + 
  theme_minimal_grid() + 
  coord_flip() +
  theme(
    axis.text.x = element_text(size = 20, face = "bold", angle = 45, vjust = 0.5),
    axis.text.y = element_text(size = 30, face = "bold"),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold"), 
    panel.spacing = unit(4, "lines")
  )
p4 

dev.off()


```

#### Supplemental Figure -- LFC correlation between genes within the GO terms -- all GO terms
```{r}
x <- 
  read.csv("~/Desktop/NSF_modelSelection/go_overlaps_allPop_treatment_Revigo.csv", sep = ',', ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>% 
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               TmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               TmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               TmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               TmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID")

y <- 
  read.csv("~/Desktop/NSF_modelSelection/go_overlaps_allPop_phenotype_Revigo.csv", sep = ',', ) %>%
  mutate(., Representative = as.character(Representative)) %>%
  mutate(., TermID = as.character(TermID)) %>%
  mutate(Representative = ifelse(.$Representative == " null", .$TermID, .$Representative)) %>%
  mutate(Representative = gsub("GO:", "", .$Representative) %>% gsub(" ", "", .)) %>%
  mutate(Representative = formatC(as.numeric(.$Representative), width = 7, format = "d", flag = "0")) %>% 
  mutate(Representative = gsub("^", "GO:", as.character(.$Representative))) %>%
  rename(TermID = "GO.ID", Representative = "Grouped_GO.ID") %>% 
  dplyr::select(., c(GO.ID, Grouped_GO.ID)) %>%
  inner_join(., 
             rbind(
               TmainLFC_guiana_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Guiana", nrow(.))),
               TmainLFC_iquitos_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Iquitos", nrow(.))),
               TmainLFC_maranon_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Marañón", nrow(.))),
               TmainLFC_nanay_lfc1000_slimGOTerms_LFC %>% 
                 add_column(., Population = rep("Nanay", nrow(.)))
             ),
             by = "GO.ID")

go_correlations_exprTrt <- data.frame()
go_correlations_exprPhe <- data.frame()

# Treatment
for (t in unique(x$Term)){
  tmp <- 
  x %>%
  subset(., Term == t) %>%
  dplyr::select(., Gene) %>%
  unique() %>%
  inner_join(., norm_expr_trt, by = "Gene") %>%
  #pivot_wider(id_cols = Gene, names_from = Population, values_from = log2FoldChange) %>%
  column_to_rownames(., var = "Gene") %>%
  cor(., method = "spearman") 
  
  tmp[upper.tri(tmp)] <- NA
  
  tmp <- as.numeric(na.omit(as.numeric(tmp))) %>%
    .[.<1.000]
  
  go_correlations_exprTrt <-
    rbind(go_correlations_exprTrt, 
          data.frame(Corr = tmp) %>%
            add_column(Term = rep(t, nrow(.)),
            Contrast = rep("Treatment", nrow(.)))
          )
}

# Phenotype
for (t in unique(y$Term)){
  tmp <- 
  y %>%
  subset(., Term == t) %>%
  dplyr::select(., Gene) %>%
  unique() %>%
  inner_join(., norm_expr_trt, by = "Gene") %>%
  #pivot_wider(id_cols = Gene, names_from = Population, values_from = log2FoldChange) %>%
  column_to_rownames(., var = "Gene") %>%
  cor(., method = "spearman") 
  
  tmp[upper.tri(tmp)] <- NA
  
  tmp <- as.numeric(na.omit(as.numeric(tmp))) %>%
    .[.<1.000]
  
  go_correlations_exprPhe <-
    rbind(go_correlations_exprPhe, 
          data.frame(Corr = tmp) %>%
            add_column(Term = rep(t, nrow(.)),
            Contrast = rep("Treatment", nrow(.)))
          )
}


pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/Fig6_exprCorr_goTerms.pdf",
    height = 40,
    width = 48)

ggplot(go_correlations_exprTrt, aes(x = Term, y = Corr)) +
  geom_point(size = 10) + 
  stat_summary(fun = mean, color = "blue", shape = 18, size = 5) +
  labs( title = "Treatment", y = "Correlation Coefficient (r)", x = NULL) +
  scale_y_continuous(breaks = seq(-1, 1, 0.1)) + 
  theme_minimal_grid() + 
  coord_flip() +
  theme(
    axis.text.x = element_text(size = 20, face = "bold", angle = 45, vjust = 0.6),
    axis.text.y = element_text(size = 30, face = "bold"),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold"), 
  )

ggplot(go_correlations_exprPhe, aes(x = Term, y = Corr)) +
  geom_point(size = 10) + 
  stat_summary(fun = mean, color = "blue", shape = 18, size = 5) +
  labs(title = "Phenotype", y = "Correlation Coefficient (r)", x = NULL) +
  scale_y_continuous(breaks = seq(-1, 1, 0.1)) + 
  theme_minimal_grid() + 
  coord_flip() +
  theme(
    axis.text.x = element_text(size = 20, face = "bold", angle = 45, vjust = 0.6),
    axis.text.y = element_text(size = 30, face = "bold"),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"),
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold"), 
  )


dev.off()

```

#### Supplemental Figure --  Heatmap of most variable genes, ordered by LFC
```{r}
trt_go_effect <- subset(TMainLFC, padj < 0.05) %>%
  dplyr::select(., c(Gene, log2FoldChange)) %>%
  arrange(desc(abs(log2FoldChange))) %>%
  head(., n = 100) %>%
  inner_join(., TmainLFC_guiana, by = "Gene") %>%
  inner_join(., TmainLFC_iquitos, by = "Gene") %>%
  inner_join(., TmainLFC_maranon, by = "Gene") %>%
  inner_join(., TmainLFC_nanay, by = "Gene") %>%
  .[,c(grep("Gene", colnames(.)),
       grep("log2FoldChange", colnames(.)))] %>%
  setNames(., nm = c("Gene", "Full", "Guiana", "Iquitos", "Marañón", "Nanay")) %>%
  melt(., variable.name = "Population", value.name = "log2FoldChange") %>%
  mutate(Contrast = rep("Treatment", nrow(.)))

pheno_go_effect <- subset(PMainLFC, padj < 0.05) %>%
  dplyr::select(., c(Gene, log2FoldChange)) %>%
  arrange(desc(abs(log2FoldChange))) %>%
  head(., n = 100) %>%
  inner_join(., PmainLFC_guiana, by = "Gene") %>%
  inner_join(., PmainLFC_iquitos, by = "Gene") %>%
  inner_join(., PmainLFC_maranon, by = "Gene") %>%
  inner_join(., PmainLFC_nanay, by = "Gene") %>%
  .[,c(grep("Gene", colnames(.)),
       grep("log2FoldChange", colnames(.)))] %>%
  setNames(., nm = c("Gene", "Full", "Guiana", "Iquitos", "Marañón", "Nanay")) %>%
  melt(., variable.name = "Population", value.name = "log2FoldChange") %>%
  mutate(Contrast = rep("Phenotype", nrow(.)))

```

## Calculations and plots for figure displaying divergent expression profiles across groups
#### Supplemental Figure --  Prop of overlapping genes for various sized LFC cross-sections
```{r}
# Calculate the # population-specific vs shared genes for different sized cross sections of the data.
# Is the lack of overlap we observe in top ranked genes due to change? 
# Take home: There are a lot of genes that are private to each genetic group,
# But there are still fewer (i.e. more is shared) than we would expect from random pulls

# Function to subset the top N genes, ranked by LFC
fTopN <- 
  function(x, n){
  x$abs <- abs(x$log2FoldChange)
  x <- x[order(x$abs, decreasing = TRUE),] %>% 
    .[1:n,] %>% 
    .$Gene
  x <- as.character(x)
  }

# Function to subset a random set of N genes from a data frame
randTopN <- function(x, n){
  x <- x %>% 
    slice_sample(n = n) %>% 
    .$Gene %>% 
    as.character(.)
}

########################
### TREATMENT -- LFC ###
########################

# Treatment -- Pull top N genes ranked LFC
genetic_group <- c("Guiana","Iquitos","Marañón","Nanay")
subsections <- seq(0, 2000, by = 200)[2:11]
propUniq_lfcCrossSections <- data.frame()
for (i in subsections){
  
  tmp <- attr(venn(list(
        Guiana = fTopN(subset(TmainLFC_all_pop, Population == "Guiana"), i),
        Iquitos = fTopN(subset(TmainLFC_all_pop, Population == "Iquitos"), i),
        Marañón = fTopN(subset(TmainLFC_all_pop, Population == "Marañón"), i),
        Nanay = fTopN(subset(TmainLFC_all_pop, Population == "Nanay"), i)), 
        show.plot = FALSE),"intersections")
  
  
  for (k in genetic_group){
    propUniq_lfcCrossSections <- 
      rbind(propUniq_lfcCrossSections, data.frame(
      group = k,
      total = i,
      num_private = length(tmp[[k]]),
      prop_private = length(tmp[[k]]) / i
    ))
  }
}
propUniq_lfcCrossSections$sample <- rep("LFC", nrow(propUniq_lfcCrossSections))


# Treatment -- Pull N genes at random
propUniq_randCrossSections <- data.frame()
for (i in subsections){
  
  tmp <- attr(venn(list(
        Guiana = randTopN(subset(TmainLFC_all_pop, Population == "Guiana"), i),
        Iquitos = randTopN(subset(TmainLFC_all_pop, Population == "Iquitos"), i),
        Marañón = randTopN(subset(TmainLFC_all_pop, Population == "Marañón"), i),
        Nanay = randTopN(subset(TmainLFC_all_pop, Population == "Nanay"), i)), 
        show.plot = FALSE),"intersections")
  
  
  for (k in genetic_group){
    propUniq_randCrossSections <- 
      rbind(propUniq_randCrossSections, data.frame(
      group = k,
      total = i,
      num_private = length(tmp[[k]]),
      prop_private = length(tmp[[k]]) / i
    ))
  }
}
propUniq_randCrossSections$sample <- rep("Random", nrow(propUniq_randCrossSections))

# Mean and stdev for this subset
propUniq_lfcCrossSections %>% 
  group_by(total) %>% 
  summarise(mPropPriv = mean(prop_private)) %>%
  ungroup() %>% 
  summarise(mean(mPropPriv),
            sd(mPropPriv))


overlap_crossSections <- rbind(propUniq_lfcCrossSections, propUniq_randCrossSections)
ggplot(overlap_crossSections, aes(x = total, y = prop_private, color = sample)) + 
  geom_point() + 
  theme_minimal_grid() +
  #theme_bw() + 
  scale_color_manual(values = c("darkblue", "orange")) + 
  labs(x = "Sample Size", 
       y = "Proportion Unique to Each Genetic Group", 
       color = NULL) + 
  stat_summary(fun = mean, shape = 3) +
  theme(
    legend.position = "top", 
    legend.justification = "center", 
    legend.direction = "horizontal"
  )

ggsave("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/overlap_lfcCrossSections_propUniqTreatment.pdf",
       height = 4.5, 
       width = 4.5)

overlap_crossSections <- 
  overlap_crossSections %>%
  mutate(total = as.character(total))
mod <- aov( prop_private ~ total + sample + total:sample, overlap_crossSections )
summary(mod)
TukeyHSD(mod, method = "fdr")


########################
### PHENOTYPE -- LFC ###
########################


# Phenotype -- Pull top N genes ranked LFC
genetic_group <- c("Guiana","Iquitos","Marañón","Nanay")
subsections <- seq(0, 2000, by = 200)[2:11]
propUniq_lfcCrossSections <- data.frame()
for (i in subsections){

  tmp <- attr(venn(list(
        Guiana = fTopN(subset(PmainLFC_all_pop, Population == "Guiana"), i),
        Iquitos = fTopN(subset(PmainLFC_all_pop, Population == "Iquitos"), i),
        Marañón = fTopN(subset(PmainLFC_all_pop, Population == "Marañón"), i),
        Nanay = fTopN(subset(PmainLFC_all_pop, Population == "Nanay"), i)), 
        show.plot = FALSE),"intersections")
  
  
  for (k in genetic_group){
    propUniq_lfcCrossSections <- 
      rbind(propUniq_lfcCrossSections, data.frame(
      group = k,
      total = i,
      num_private = length(tmp[[k]]),
      prop_private = length(tmp[[k]]) / i
    ))
  }
}
propUniq_lfcCrossSections$sample <- rep("LFC", nrow(propUniq_lfcCrossSections))


# Phenotype -- Pull N genes at random
propUniq_randCrossSections <- data.frame()
for (i in subsections){
  
  tmp <- attr(venn(list(
        Guiana = randTopN(subset(PmainLFC_all_pop, Population == "Guiana"), i),
        Iquitos = randTopN(subset(PmainLFC_all_pop, Population == "Iquitos"), i),
        Marañón = randTopN(subset(PmainLFC_all_pop, Population == "Marañón"), i),
        Nanay = randTopN(subset(PmainLFC_all_pop, Population == "Nanay"), i)), 
        show.plot = FALSE),"intersections")
  
  
  for (k in genetic_group){
    propUniq_randCrossSections <- 
      rbind(propUniq_randCrossSections, data.frame(
      group = k,
      total = i,
      num_private = length(tmp[[k]]),
      prop_private = length(tmp[[k]]) / i
    ))
  }
}
propUniq_randCrossSections$sample <- rep("Random", nrow(propUniq_randCrossSections))


overlap_crossSections <- rbind(propUniq_lfcCrossSections, propUniq_randCrossSections)
overlap_crossSections %>% 
  ggplot(aes(x = total, y = prop_private, color = sample)) + 
  geom_point() + 
  theme_minimal_grid() +
  scale_color_manual(values = c("darkblue", "orange")) + 
  labs(x = "Sample Size", 
       y = "Proportion Unique to Each Genetic Group", 
       color = NULL) + 
  stat_summary(fun = mean, shape = 3) +
  theme(
    legend.position = "top", 
    legend.justification = "center", 
    legend.direction = "horizontal"
  )

ggsave("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/overlap_lfcCrossSections_propUniqPhenotype.pdf",
       height = 4.5, 
       width = 4.5)


#############################
### Treatment -- P-Values ###
#############################

tmp <- attr(venn(list(
        Guiana = subset(TmainLFC_all_pop, Population == "Guiana" & padj < 0.05)$Gene,
        Iquitos = subset(TmainLFC_all_pop, Population == "Iquitos" & padj < 0.05)$Gene,
        Marañón = subset(TmainLFC_all_pop, Population == "Marañón" & padj < 0.05)$Gene,
        Nanay = subset(TmainLFC_all_pop, Population == "Nanay" & padj < 0.05)$Gene),
        show.plot = FALSE),"intersections")
  

tmp_df <- data.frame()
for (k in genetic_group){
    tmp_df <- 
      rbind(tmp_df, data.frame(
      group = k,
      total = length(as.character(subset(TmainLFC_all_pop, Population == k & padj < 0.05)$Gene)),
      num_private = length(tmp[[k]]),
      prop_private = length(tmp[[k]]) / length(subset(TmainLFC_all_pop, Population == k & padj < 0.05)$Gene)
    ))
}

tmp_df %>% 
  summarise(mean(prop_private),
            sd(prop_private))


#############################
### Phenotype -- P-Values ###
#############################

tmp <- attr(venn(list(
        Guiana = subset(PmainLFC_all_pop, Population == "Guiana" & padj < 0.05)$Gene,
        Iquitos = subset(PmainLFC_all_pop, Population == "Iquitos" & padj < 0.05)$Gene,
        Marañón = subset(PmainLFC_all_pop, Population == "Marañón" & padj < 0.05)$Gene,
        Nanay = subset(PmainLFC_all_pop, Population == "Nanay" & padj < 0.05)$Gene),
        show.plot = FALSE),"intersections")
  

tmp_df <- data.frame()
for (k in genetic_group){
    tmp_df <- 
      rbind(tmp_df, data.frame(
      group = k,
      total = length(as.character(subset(PmainLFC_all_pop, Population == k & padj < 0.05)$Gene)),
      num_private = length(tmp[[k]]),
      prop_private = length(tmp[[k]]) / length(subset(PmainLFC_all_pop, Population == k & padj < 0.05)$Gene)
    ))
}

tmp_df %>% 
  summarise(mean(prop_private),
            sd(prop_private))



```

#### Supplemental Figure --  Gene expression for private vs shared genes, top 1000 genes by | LFC |
```{r}
# Create dataframes
norm_expr_grpMeans <- norm_expr %>%
  t() %>%
  merge(dplyr::select(colData, Gen_Group, Treatment), ., by = 0) %>% 
  column_to_rownames(var = "Row.names") %>%
  pivot_longer(cols = starts_with("SCA-6"), names_to = "Gene", values_to = "Expression") %>%
  group_by(Gen_Group, Gene) %>% 
  summarise(mExpr = mean(Expression))

privateVsShared <- treat_df %>%
  rownames_to_column(var = "Gene") %>% 
  mutate(gen_grp_shared = as.character(paste0(Guiana, Iquitos, Marañón, Nanay)))%>% 
  mutate(gen_grp_shared = ifelse(gen_grp_shared == "1000", "Guiana_private",
                                 ifelse(gen_grp_shared == "0100", "Iquitos_private",
                                        ifelse(gen_grp_shared == "0010", "Marañón_private",
                                               ifelse(gen_grp_shared == "0001", "Nanay_private", "shared"))))) %>%
  inner_join(norm_expr_grpMeans, by = "Gene") %>%
  mutate(shared = ifelse(gen_grp_shared == "shared", "shared", "unique")) %>% 
  mutate(gen_grp_shared = ifelse(gen_grp_shared == "shared", 
                                 paste0(Gen_Group, "_shared"),
                                 gen_grp_shared))
  
# Plot expression for private vs shared genes
pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/private_vs_shared_expression_treat.pdf", 
    width = 4.5,
    height = 4.5)
dodge <- position_dodge(width = 1)
ggplot(privateVsShared, aes(x = gen_grp_shared, y = mExpr, fill = shared)) +
  geom_violin(position = dodge) +
  geom_boxplot(position = dodge, outlier.color = NA, 
               width = 0.1, show.legend = FALSE) + 
  theme_bw() + 
  labs(x = NULL, y = "Normalized Expression (VST)", fill = NULL)+
  theme(
    legend.position = "top", 
    legend.justification = "center", 
    legend.direction = "horizontal",
    axis.text = element_text(size = 10), 
    legend.text = element_text(size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title.y = element_text(size = 10)
  ) + 
  scale_x_discrete(labels = c("Guiana \nUnique", "Guiana \nShared",
                              "Iquitos \nUnique", "Iquitos \nShared",
                              "Marañón \nUnique", "Marañón \nShared",
                              "Nanay \nUnique", "Nanay \nShared"))
dev.off()


# Are the observed differences significant?
mod <- aov(mExpr ~ gen_grp_shared, data = privateVsShared)
summary(mod)
TukeyHSD(mod, method = "fdr")



# Phenotype
privateVsShared <- pheno_df %>%
  rownames_to_column(var = "Gene") %>% 
  mutate(gen_grp_shared = as.character(paste0(Guiana, Iquitos, Marañón, Nanay)))%>% 
  mutate(gen_grp_shared = ifelse(gen_grp_shared == "1000", "Guiana_private",
                                 ifelse(gen_grp_shared == "0100", "Iquitos_private",
                                        ifelse(gen_grp_shared == "0010", "Marañón_private",
                                               ifelse(gen_grp_shared == "0001", "Nanay_private", "shared"))))) %>%
  inner_join(norm_expr_grpMeans, by = "Gene") %>%
  mutate(shared = ifelse(gen_grp_shared == "shared", "shared", "unique")) %>% 
  mutate(gen_grp_shared = ifelse(gen_grp_shared == "shared", 
                                 paste0(Gen_Group, "_shared"),
                                 gen_grp_shared))
  
# Plot expression for private vs shared genes
pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/private_vs_shared_expression_pheno.pdf", 
    width = 4.5,
    height = 4.5)
dodge <- position_dodge(width = 1)
ggplot(privateVsShared, aes(x = gen_grp_shared, y = mExpr, fill = shared)) +
  geom_violin(position = dodge) +
  geom_boxplot(position = dodge, outlier.color = NA, 
               width = 0.1, show.legend = FALSE) + 
  theme_bw() + 
  labs(x = NULL, y = "Normalized Expression (VST)", fill = NULL)+
  theme(
    legend.position = "top", 
    legend.justification = "center", 
    legend.direction = "horizontal",
    axis.text = element_text(size = 10), 
    legend.text = element_text(size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title.y = element_text(size = 10)
  ) + 
  scale_x_discrete(labels = c("Guiana \nUnique", "Guiana \nShared",
                              "Iquitos \nUnique", "Iquitos \nShared",
                              "Marañón \nUnique", "Marañón \nShared",
                              "Nanay \nUnique", "Nanay \nShared"))
dev.off()



# Are the observed differences significant?
mod <- aov(mExpr ~ gen_grp_shared, data = privateVsShared)
summary(mod)
TukeyHSD(mod, method = "fdr")

```

#### Supplemental figure, upset overlap for each genetic group--clustered paralogs
```{r}
sca6_cluster <- read.table("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Theobroma_cacao_SCA-6_chr.v1.0.fractional.counts.tsv", header = TRUE) %>% 
  dplyr::select(c(GeneID, ClusterID)) %>% 
  setNames(nm = c("Gene", "Cluster")) 


sca6_cluster_lfc1000_trt <- inner_join(sca6_cluster, TmainLFC_all_pop_lfc1000, by = "Gene")
lst_trt <-
  list(
  Guiana = as.character(subset(sca6_cluster_lfc1000_trt, Population == "Guiana")$Cluster),
  Iquitos = as.character(subset(sca6_cluster_lfc1000_trt, Population == "Iquitos")$Cluster),
  Marañón = as.character(subset(sca6_cluster_lfc1000_trt, Population == "Marañón")$Cluster),
  Nanay = as.character(subset(sca6_cluster_lfc1000_trt, Population == "Nanay")$Cluster)
)
lst_trt <- 
  list_to_matrix(lst_trt) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

sca6_cluster_lfc1000_phe <- inner_join(sca6_cluster, PmainLFC_all_pop_lfc1000, by = "Gene")
lst_phe <-
  list(
  Guiana = as.character(subset(sca6_cluster_lfc1000_phe, Population == "Guiana")$Cluster),
  Iquitos = as.character(subset(sca6_cluster_lfc1000_phe, Population == "Iquitos")$Cluster),
  Marañón = as.character(subset(sca6_cluster_lfc1000_phe, Population == "Marañón")$Cluster),
  Nanay = as.character(subset(sca6_cluster_lfc1000_phe, Population == "Nanay")$Cluster)
)
lst_phe <- 
  list_to_matrix(lst_phe) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

sca6_cluster_lfc1000_add <- inner_join(sca6_cluster_lfc1000_trt, sca6_cluster_lfc1000_phe, by = c("Population", "Gene"))
lst_add <-
  list(
  Guiana = as.character(subset(sca6_cluster_lfc1000_add, Population == "Guiana")$Cluster.x),
  Iquitos = as.character(subset(sca6_cluster_lfc1000_add, Population == "Iquitos")$Cluster.x),
  Marañón = as.character(subset(sca6_cluster_lfc1000_add, Population == "Marañón")$Cluster.x),
  Nanay = as.character(subset(sca6_cluster_lfc1000_add, Population == "Nanay")$Cluster.x)
)
lst_add <- 
  list_to_matrix(lst_add) %>%
  data.frame(.) %>%
  setNames(., nm = c("Guiana", "Iquitos", "Marañón", "Nanay"))

pdf("/Users/noah/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/upsetPlots_degOverlaps_paralogs.pdf", width = 12)

 # Treatment
 upset(lst_trt, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Genes (Largest |LFC|)",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))
 
 # Phenotype
 upset(lst_phe, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Genes (Largest |LFC|)",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))
 
 # Additive
 upset(lst_add, main.bar.color = "black",
      queries = list(
  #list(query = intersects, params = list("Full"), color = "black", active = T),
  list(query = intersects, params = list("Guiana"), color = "steelblue", active = T),
  list(query = intersects, params = list("Iquitos"), color = "firebrick", active = T),
  list(query = intersects, params = list("Marañón"), color = "darkgreen", active = T),
  list(query = intersects, params = list("Nanay"), color = "darkorange", active = T),
  list(query = intersects, params = list("Guiana", "Iquitos", "Marañón", "Nanay"), color = "#Df5286", active = T)),
  #sets.bar.color = c("black","darkorange","darkgreen","firebrick", "steelblue"),
  keep.order = TRUE,
  sets.x.label = "Genes (Largest |LFC|)",
  mainbar.y.label = "Genes in Each Intersection",
  point.size = 4.5,
  line.size = 3, 
  text.scale = c(2.5, 2.5, 2, 2, 2.5, 2))
dev.off()


```

#### Suplemental figure, plot for unbalanced design
```{r}
plot_df <- colData %>%
  subset(., Gen_Group != "Contamana") %>%
  subset(., Gen_Group != "Hybrid") %>% 
  group_by(Genotype) %>% 
  tally() %>%
  inner_join(colData, by = "Genotype") %>%
  dplyr::select(., c(Genotype, Gen_Group, Phenotype, n)) %>%
  distinct() %>%
  arrange(Gen_Group) %>%
  mutate(Gen_Group = gsub("Maranon", "Marañón", .$Gen_Group))

p1 <- 
  ggplot(plot_df, aes(x = interaction(Genotype, Phenotype, Gen_Group, sep = "-"), y = n, fill = Gen_Group)) +
  geom_histogram(stat = "identity") + 
  geom_text( aes(label = n), vjust = -0.1, size = 12, fontface = "bold") +
  theme_minimal_grid() + 
  scale_fill_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  #scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9")) +
  labs(x = "Genotype", 
       y = "# of Samples") +
  theme(
    axis.text.x = element_text(size = 30, angle = 90, vjust = 0.5),
    axis.text.y = element_text(size = 30),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"), 
    legend.position = "none"
    )

# Totals for label
p2 <- 
  plot_df %>%
  group_by(Gen_Group, Phenotype) %>%
  summarise(samples = sum(n)) %>%
  ggplot(., aes(x = Phenotype, y = samples, fill = Gen_Group)) +
  geom_bar(position = "dodge", stat = "identity") + 
  geom_text(aes(label = samples), vjust = -0.1, size = 12, fontface = "bold") +
  facet_wrap(~Gen_Group) +
  scale_fill_manual(values = c("steelblue", "firebrick", "darkgreen", "darkorange")) + 
  theme_minimal_grid() +
  labs(x = NULL, y = NULL) +
  theme(
    axis.text.x = element_text(size = 30, vjust = 0.5),
    axis.text.y = element_text(size = 30),
    axis.title.x = element_text(size = 30, face = "bold"),
    axis.title.y = element_text(size = 30, face = "bold"), 
    legend.text = element_text(size = 30),
    legend.title = element_text(size = 30, face = "bold"),
    legend.position = "none", 
    strip.background = element_rect(color = "lightgrey"),
    strip.text = element_text(size = 30, face = "bold")
    )

pdf("~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/unbalanced_design.pdf", useDingbats=FALSE, width = 30, height = 20)
plot_grid(p1, p2)
dev.off()
```


# Write out supplemental tables
```{r}
TmainLFC_all_pop_lfc1000 %>%
  inner_join(annot, by = "Gene") %>%
  write.table(., 
              file = "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/TableS1_TreatmentLFCTop1000.tsv",
              sep = '\t',
              quote = FALSE, 
              row.names = FALSE)

PmainLFC_all_pop_lfc1000 %>%
  inner_join(annot, by = "Gene") %>%
  write.table(., 
              file = "~/GradSchool/Research/Projects/NSF_PlantGenome/Trxome_Exp/Outline_June2021/Supplemental/TableS1_PhenotypeLFCTop1000.tsv",
              sep = '\t',
              quote = FALSE, 
              row.names = FALSE)
```



